// etherspot/dist/sdk/exchange/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiFiBridge = exports.CrossChainServiceProvider = exports.SocketTokenDirection = exports.ExchangeProviders = void 0;
var ExchangeProviders;
(function (ExchangeProviders) {
    ExchangeProviders["Uniswap"] = "Uniswap";
    ExchangeProviders["OneInch"] = "OneInch";
    ExchangeProviders["Synthetix"] = "Synthetix";
    ExchangeProviders["Sushiswap"] = "Sushiswap";
    ExchangeProviders["Honeyswap"] = "Honeyswap";
    ExchangeProviders["Paraswap"] = "Paraswap";
})(ExchangeProviders = exports.ExchangeProviders || (exports.ExchangeProviders = {}));
var SocketTokenDirection;
(function (SocketTokenDirection) {
    SocketTokenDirection["From"] = "From";
    SocketTokenDirection["To"] = "To";
})(SocketTokenDirection = exports.SocketTokenDirection || (exports.SocketTokenDirection = {}));
var CrossChainServiceProvider;
(function (CrossChainServiceProvider) {
    CrossChainServiceProvider["SocketV2"] = "SocketV2";
    CrossChainServiceProvider["LiFi"] = "LiFi";
    CrossChainServiceProvider["Connext"] = "Connext";
})(CrossChainServiceProvider = exports.CrossChainServiceProvider || (exports.CrossChainServiceProvider = {}));
var LiFiBridge;
(function (LiFiBridge) {
    LiFiBridge["across"] = "across";
    LiFiBridge["arbitrum"] = "arbitrum";
    LiFiBridge["avalanche"] = "avalanche";
    LiFiBridge["cbridge"] = "cbridge";
    LiFiBridge["connext"] = "connext";
    LiFiBridge["hop"] = "hop";
    LiFiBridge["hyphen"] = "hyphen";
    LiFiBridge["multichain"] = "multichain";
    LiFiBridge["optimism"] = "optimism";
    LiFiBridge["polygon"] = "polygon";
    LiFiBridge["stargate"] = "stargate";
})(LiFiBridge = exports.LiFiBridge || (exports.LiFiBridge = {}));
