// etherspot/dist/sdk/rates/rates.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RatesService = void 0;
const core_1 = require("@apollo/client/core");
const common_1 = require("../common");
const classes_1 = require("./classes");
class RatesService extends common_1.Service {
    async fetchExchangeRates(tokens, ChainId) {
        const { apiService } = this.services;
        const { result } = await apiService.query(core_1.gql `
        query($tokens: [String!]!, $ChainId: Int!) {
          result: fetchExchangeRates(tokens: $tokens, chainId: $ChainId) {
            errored
            error
            items {
              address
              eth
              eur
              gbp
              usd
            }
          }
        }
      `, {
            variables: {
                tokens,
                ChainId,
            },
            models: {
                result: classes_1.RateData,
            },
        });
        return result;
    }
}
exports.RatesService = RatesService;
