// @requestnetwork/data-access/dist/combined-data-access.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CombinedDataAccess = void 0;
const tslib_1 = require("tslib");
class CombinedDataAccess {
    constructor(reader, writer) {
        this.reader = reader;
        this.writer = writer;
        this.getTransactionsByChannelId = this.reader.getTransactionsByChannelId.bind(this.reader);
        this.getChannelsByTopic = this.reader.getChannelsByTopic.bind(this.reader);
        this.getChannelsByMultipleTopics = this.reader.getChannelsByMultipleTopics.bind(this.reader);
        this.persistTransaction = this.writer.persistTransaction.bind(this.writer);
    }
    initialize() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.reader.initialize();
            yield this.writer.initialize();
        });
    }
    close() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.writer.close();
            yield this.reader.close();
        });
    }
}
exports.CombinedDataAccess = CombinedDataAccess;
