// etherspot/dist/sdk/name-resolution/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NameResolutionRootNodeStates = exports.NameResolutionNodeZones = exports.NameResolutionNodeStates = void 0;
var NameResolutionNodeStates;
(function (NameResolutionNodeStates) {
    NameResolutionNodeStates["Reserved"] = "Reserved";
    NameResolutionNodeStates["Claimed"] = "Claimed";
})(NameResolutionNodeStates = exports.NameResolutionNodeStates || (exports.NameResolutionNodeStates = {}));
var NameResolutionNodeZones;
(function (NameResolutionNodeZones) {
    NameResolutionNodeZones["ENS"] = "ENS";
    NameResolutionNodeZones["UNS"] = "UNS";
    NameResolutionNodeZones["ZNS"] = "ZNS";
    NameResolutionNodeZones["FIO"] = "FIO";
})(NameResolutionNodeZones = exports.NameResolutionNodeZones || (exports.NameResolutionNodeZones = {}));
var NameResolutionRootNodeStates;
(function (NameResolutionRootNodeStates) {
    NameResolutionRootNodeStates["Submitted"] = "Submitted";
    NameResolutionRootNodeStates["Verified"] = "Verified";
    NameResolutionRootNodeStates["Released"] = "Released";
})(NameResolutionRootNodeStates = exports.NameResolutionRootNodeStates || (exports.NameResolutionRootNodeStates = {}));
