// etherspot/dist/sdk/name-resolution/name-resolution.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NameResolutionService = void 0;
const core_1 = require("@apollo/client/core");
const common_1 = require("../common");
const classes_1 = require("./classes");
class NameResolutionService extends common_1.Service {
    async resolveName(chainId, name) {
        const { apiService } = this.services;
        const { result } = await apiService.query(core_1.gql `
        query($chainId: Int, $name: String!) {
          result: resolveName(chainId: $chainId, name: $name) {
            results {
              ens {
                address
                hash
                name
                label
                state
                zone
              }
              unstoppabledomains {
                address
                hash
                name
                label
                state
                zone
              }
              fio {
                address
                hash
                name
                label
                state
                zone
                fioChainCode
                fioTokenCode
              }
            }
            message
            failed
          }
        }
      `, {
            variables: {
                chainId,
                name,
            },
            models: {
                result: classes_1.NameResolutionsNodes,
            },
        });
        return result;
    }
}
exports.NameResolutionService = NameResolutionService;
