// @requestnetwork/request-logic/dist/actions/accept.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@requestnetwork/types");
const action_1 = require("../action");
const request_1 = require("../request");
const version_1 = require("../version");
const utils_1 = require("@requestnetwork/utils");
/**
 * Implementation of the action accept from request logic specification
 */
exports.default = {
    applyActionToRequest,
    format,
};
/**
 * Function to format an action to accept a Request
 *
 * @param acceptParameters parameters to accept a request
 * @param IIdentity Identity of the signer
 * @param ISignatureProvider Signature provider in charge of the signature
 *
 * @returns IAction  the action with the signature
 */
function format(acceptParameters, signerIdentity, signatureProvider) {
    const unsignedAction = {
        name: types_1.RequestLogicTypes.ACTION_NAME.ACCEPT,
        parameters: acceptParameters,
        version: version_1.default.currentVersion,
    };
    return action_1.default.createAction(unsignedAction, signerIdentity, signatureProvider);
}
/**
 * Function to apply an Accept action on a request
 *
 * @param Types.IAction action  the action to apply
 *
 * @returns Types.IRequest the new request
 */
function applyActionToRequest(action, timestamp, request) {
    if (!action.data.parameters.requestId) {
        throw new Error('requestId must be given');
    }
    if (!request.payer) {
        throw new Error('the request must have a payer');
    }
    if (request.state !== types_1.RequestLogicTypes.STATE.CREATED) {
        throw new Error('the request state must be created');
    }
    const signer = action_1.default.getSignerIdentityFromAction(action);
    const signerRole = request_1.default.getRoleInRequest(signer, request);
    if (signerRole === types_1.RequestLogicTypes.ROLE.PAYER) {
        request.state = types_1.RequestLogicTypes.STATE.ACCEPTED;
    }
    else {
        throw new Error('Signer must be the payer');
    }
    // avoid to mutate the request
    let requestCopied = (0, utils_1.deepCopy)(request);
    requestCopied = request_1.default.pushExtensionsData(requestCopied, action.data.parameters.extensionsData);
    requestCopied.events.push(generateEvent(action, timestamp, signer));
    return requestCopied;
}
/**
 * Private function to generate the event 'Accept' from an action
 *
 * @param Types.IAction action the action that create the event
 * @param IdentityTypes.IIdentity actionSigner the signer of the action
 *
 * @returns Types.IEvent the event generated
 */
function generateEvent(action, timestamp, actionSigner) {
    const params = action.data.parameters;
    const event = {
        actionSigner,
        name: types_1.RequestLogicTypes.ACTION_NAME.ACCEPT,
        parameters: {
            extensionsDataLength: params.extensionsData ? params.extensionsData.length : 0,
        },
        timestamp,
    };
    return event;
}
