// @requestnetwork/request-logic/dist/actions/addStakeholders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@requestnetwork/types");
const utils_1 = require("@requestnetwork/utils");
const action_1 = require("../action");
const request_1 = require("../request");
const version_1 = require("../version");
/**
 * Implementation of the action add stakeholder from request logic specification
 */
exports.default = {
    applyActionToRequest,
    format,
};
/**
 * Function to format an action to add stakeholders to a Request
 *
 * @param IAddStakeholdersParameters addStakeholdersParameters parameters to add stakeholders to a request
 * @param IIdentity signerIdentity Identity of the signer
 * @param ISignatureProvider signatureProvider Signature provider in charge of the signature
 *
 * @returns IAction  the action with the signature
 */
function format(addStakeholderParameters, signerIdentity, signatureProvider) {
    const unsignedAction = {
        name: types_1.RequestLogicTypes.ACTION_NAME.ADD_STAKEHOLDERS,
        parameters: addStakeholderParameters,
        version: version_1.default.currentVersion,
    };
    return action_1.default.createAction(unsignedAction, signerIdentity, signatureProvider);
}
/**
 * Function to apply an addStakeholder action to a request
 *
 * @param Types.IAction action the action to apply
 *
 * @returns Types.IRequest the new request
 */
function applyActionToRequest(action, timestamp, request) {
    if (!action.data.parameters.requestId) {
        throw new Error('requestId must be given');
    }
    const signer = action_1.default.getSignerIdentityFromAction(action);
    // avoid to mutate the request
    let requestCopied = (0, utils_1.deepCopy)(request);
    requestCopied = request_1.default.pushExtensionsData(requestCopied, action.data.parameters.extensionsData);
    requestCopied.events.push(generateEvent(action, timestamp, signer));
    return requestCopied;
}
/**
 * Private function to generate the event 'addStakeholder' from an action
 *
 * @param Types.IAction action the action that create the event
 * @param IdentityTypes.IIdentity actionSigner the signer of the action
 *
 * @returns Types.IEvent the event generated
 */
function generateEvent(action, timestamp, actionSigner) {
    const params = action.data.parameters;
    const event = {
        actionSigner,
        name: types_1.RequestLogicTypes.ACTION_NAME.ADD_STAKEHOLDERS,
        parameters: {
            extensionsDataLength: params.extensionsData ? params.extensionsData.length : 0,
        },
        timestamp,
    };
    return event;
}
