// @requestnetwork/payment-detection/dist/thegraph/generated/graphql-superfluid.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuperTokenLogicCreatedEvent_OrderBy = exports.SuperTokenFactoryUpdatedEvent_OrderBy = exports.SuperTokenCreatedEvent_OrderBy = exports.SubscriptionUnitsUpdatedEvent_OrderBy = exports.SubscriptionRevokedEvent_OrderBy = exports.SubscriptionDistributionClaimedEvent_OrderBy = exports.SubscriptionApprovedEvent_OrderBy = exports.Stream_OrderBy = exports.StreamRevision_OrderBy = exports.StreamPeriod_OrderBy = exports.SetEvent_OrderBy = exports.SentEvent_OrderBy = exports.SfMeta_OrderBy = exports.RoleRevokedEvent_OrderBy = exports.RoleGrantedEvent_OrderBy = exports.RoleAdminChangedEvent_OrderBy = exports.RewardAddressChangedEvent_OrderBy = exports.ResolverEntry_OrderBy = exports.PppConfigurationChangedEvent_OrderBy = exports.OrderDirection = exports.NewPicEvent_OrderBy = exports.MintedEvent_OrderBy = exports.JailEvent_OrderBy = exports.Index_OrderBy = exports.IndexUpdatedEvent_OrderBy = exports.IndexUnsubscribedEvent_OrderBy = exports.IndexUnitsUpdatedEvent_OrderBy = exports.IndexSubscription_OrderBy = exports.IndexSubscribedEvent_OrderBy = exports.IndexDistributionClaimedEvent_OrderBy = exports.IndexCreatedEvent_OrderBy = exports.GovernanceReplacedEvent_OrderBy = exports.FlowUpdatedEvent_OrderBy = exports.FlowOperator_OrderBy = exports.FlowOperatorUpdatedEvent_OrderBy = exports.ExitRateChangedEvent_OrderBy = exports.Event_OrderBy = exports.CustomSuperTokenCreatedEvent_OrderBy = exports.ConfigChangedEvent_OrderBy = exports.CfAv1LiquidationPeriodChangedEvent_OrderBy = exports.BurnedEvent_OrderBy = exports.BondIncreasedEvent_OrderBy = exports.AppRegisteredEvent_OrderBy = exports.AgreementLiquidatedV2Event_OrderBy = exports.AgreementLiquidatedByEvent_OrderBy = exports.AgreementClassUpdatedEvent_OrderBy = exports.AgreementClassRegisteredEvent_OrderBy = exports.Account_OrderBy = exports.AccountTokenSnapshot_OrderBy = exports.AccountTokenSnapshotLog_OrderBy = void 0;
exports.GetSuperFluidEvents = exports.PaymentFragment = exports.EventFragment = exports.getSdk = exports.GetSuperFluidEventsDocument = exports.PaymentFragmentFragmentDoc = exports.EventFragmentFragmentDoc = exports._SubgraphErrorPolicy_ = exports.TrustedForwarderChangedEvent_OrderBy = exports.TransferEvent_OrderBy = exports.Token_OrderBy = exports.TokenUpgradedEvent_OrderBy = exports.TokenStatistic_OrderBy = exports.TokenStatisticLog_OrderBy = exports.TokenGovernanceConfig_OrderBy = exports.TokenDowngradedEvent_OrderBy = exports.SuperTokenMinimumDepositChangedEvent_OrderBy = exports.SuperTokenLogicUpdatedEvent_OrderBy = void 0;
const graphql_tag_1 = require("graphql-tag");
var AccountTokenSnapshotLog_OrderBy;
(function (AccountTokenSnapshotLog_OrderBy) {
    AccountTokenSnapshotLog_OrderBy["Id"] = "id";
    AccountTokenSnapshotLog_OrderBy["Timestamp"] = "timestamp";
    AccountTokenSnapshotLog_OrderBy["BlockNumber"] = "blockNumber";
    AccountTokenSnapshotLog_OrderBy["TransactionHash"] = "transactionHash";
    AccountTokenSnapshotLog_OrderBy["LogIndex"] = "logIndex";
    AccountTokenSnapshotLog_OrderBy["Order"] = "order";
    AccountTokenSnapshotLog_OrderBy["TriggeredByEventName"] = "triggeredByEventName";
    AccountTokenSnapshotLog_OrderBy["MaybeCriticalAtTimestamp"] = "maybeCriticalAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["TotalNumberOfActiveStreams"] = "totalNumberOfActiveStreams";
    AccountTokenSnapshotLog_OrderBy["ActiveOutgoingStreamCount"] = "activeOutgoingStreamCount";
    AccountTokenSnapshotLog_OrderBy["ActiveIncomingStreamCount"] = "activeIncomingStreamCount";
    AccountTokenSnapshotLog_OrderBy["TotalNumberOfClosedStreams"] = "totalNumberOfClosedStreams";
    AccountTokenSnapshotLog_OrderBy["InactiveOutgoingStreamCount"] = "inactiveOutgoingStreamCount";
    AccountTokenSnapshotLog_OrderBy["InactiveIncomingStreamCount"] = "inactiveIncomingStreamCount";
    AccountTokenSnapshotLog_OrderBy["TotalSubscriptionsWithUnits"] = "totalSubscriptionsWithUnits";
    AccountTokenSnapshotLog_OrderBy["TotalApprovedSubscriptions"] = "totalApprovedSubscriptions";
    AccountTokenSnapshotLog_OrderBy["Balance"] = "balance";
    AccountTokenSnapshotLog_OrderBy["TotalDeposit"] = "totalDeposit";
    AccountTokenSnapshotLog_OrderBy["TotalNetFlowRate"] = "totalNetFlowRate";
    AccountTokenSnapshotLog_OrderBy["TotalInflowRate"] = "totalInflowRate";
    AccountTokenSnapshotLog_OrderBy["TotalOutflowRate"] = "totalOutflowRate";
    AccountTokenSnapshotLog_OrderBy["TotalAmountStreamedIn"] = "totalAmountStreamedIn";
    AccountTokenSnapshotLog_OrderBy["TotalAmountStreamedOut"] = "totalAmountStreamedOut";
    AccountTokenSnapshotLog_OrderBy["TotalAmountStreamed"] = "totalAmountStreamed";
    AccountTokenSnapshotLog_OrderBy["TotalAmountTransferred"] = "totalAmountTransferred";
    AccountTokenSnapshotLog_OrderBy["Account"] = "account";
    AccountTokenSnapshotLog_OrderBy["Account__id"] = "account__id";
    AccountTokenSnapshotLog_OrderBy["Account__createdAtTimestamp"] = "account__createdAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["Account__createdAtBlockNumber"] = "account__createdAtBlockNumber";
    AccountTokenSnapshotLog_OrderBy["Account__updatedAtTimestamp"] = "account__updatedAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["Account__updatedAtBlockNumber"] = "account__updatedAtBlockNumber";
    AccountTokenSnapshotLog_OrderBy["Account__isSuperApp"] = "account__isSuperApp";
    AccountTokenSnapshotLog_OrderBy["Token"] = "token";
    AccountTokenSnapshotLog_OrderBy["Token__id"] = "token__id";
    AccountTokenSnapshotLog_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    AccountTokenSnapshotLog_OrderBy["Token__decimals"] = "token__decimals";
    AccountTokenSnapshotLog_OrderBy["Token__name"] = "token__name";
    AccountTokenSnapshotLog_OrderBy["Token__symbol"] = "token__symbol";
    AccountTokenSnapshotLog_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    AccountTokenSnapshotLog_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    AccountTokenSnapshotLog_OrderBy["Token__isListed"] = "token__isListed";
    AccountTokenSnapshotLog_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot"] = "accountTokenSnapshot";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__id"] = "accountTokenSnapshot__id";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__updatedAtTimestamp"] = "accountTokenSnapshot__updatedAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__updatedAtBlockNumber"] = "accountTokenSnapshot__updatedAtBlockNumber";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__isLiquidationEstimateOptimistic"] = "accountTokenSnapshot__isLiquidationEstimateOptimistic";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__maybeCriticalAtTimestamp"] = "accountTokenSnapshot__maybeCriticalAtTimestamp";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalNumberOfActiveStreams"] = "accountTokenSnapshot__totalNumberOfActiveStreams";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__activeOutgoingStreamCount"] = "accountTokenSnapshot__activeOutgoingStreamCount";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__activeIncomingStreamCount"] = "accountTokenSnapshot__activeIncomingStreamCount";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalNumberOfClosedStreams"] = "accountTokenSnapshot__totalNumberOfClosedStreams";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__inactiveOutgoingStreamCount"] = "accountTokenSnapshot__inactiveOutgoingStreamCount";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__inactiveIncomingStreamCount"] = "accountTokenSnapshot__inactiveIncomingStreamCount";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalSubscriptionsWithUnits"] = "accountTokenSnapshot__totalSubscriptionsWithUnits";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalApprovedSubscriptions"] = "accountTokenSnapshot__totalApprovedSubscriptions";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__balanceUntilUpdatedAt"] = "accountTokenSnapshot__balanceUntilUpdatedAt";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalDeposit"] = "accountTokenSnapshot__totalDeposit";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalNetFlowRate"] = "accountTokenSnapshot__totalNetFlowRate";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalInflowRate"] = "accountTokenSnapshot__totalInflowRate";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalOutflowRate"] = "accountTokenSnapshot__totalOutflowRate";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalAmountStreamedInUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedInUntilUpdatedAt";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalAmountStreamedOutUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedOutUntilUpdatedAt";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalAmountStreamedUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedUntilUpdatedAt";
    AccountTokenSnapshotLog_OrderBy["AccountTokenSnapshot__totalAmountTransferredUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountTransferredUntilUpdatedAt";
})(AccountTokenSnapshotLog_OrderBy = exports.AccountTokenSnapshotLog_OrderBy || (exports.AccountTokenSnapshotLog_OrderBy = {}));
var AccountTokenSnapshot_OrderBy;
(function (AccountTokenSnapshot_OrderBy) {
    AccountTokenSnapshot_OrderBy["Id"] = "id";
    AccountTokenSnapshot_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    AccountTokenSnapshot_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    AccountTokenSnapshot_OrderBy["IsLiquidationEstimateOptimistic"] = "isLiquidationEstimateOptimistic";
    AccountTokenSnapshot_OrderBy["MaybeCriticalAtTimestamp"] = "maybeCriticalAtTimestamp";
    AccountTokenSnapshot_OrderBy["TotalNumberOfActiveStreams"] = "totalNumberOfActiveStreams";
    AccountTokenSnapshot_OrderBy["ActiveOutgoingStreamCount"] = "activeOutgoingStreamCount";
    AccountTokenSnapshot_OrderBy["ActiveIncomingStreamCount"] = "activeIncomingStreamCount";
    AccountTokenSnapshot_OrderBy["TotalNumberOfClosedStreams"] = "totalNumberOfClosedStreams";
    AccountTokenSnapshot_OrderBy["InactiveOutgoingStreamCount"] = "inactiveOutgoingStreamCount";
    AccountTokenSnapshot_OrderBy["InactiveIncomingStreamCount"] = "inactiveIncomingStreamCount";
    AccountTokenSnapshot_OrderBy["TotalSubscriptionsWithUnits"] = "totalSubscriptionsWithUnits";
    AccountTokenSnapshot_OrderBy["TotalApprovedSubscriptions"] = "totalApprovedSubscriptions";
    AccountTokenSnapshot_OrderBy["BalanceUntilUpdatedAt"] = "balanceUntilUpdatedAt";
    AccountTokenSnapshot_OrderBy["TotalDeposit"] = "totalDeposit";
    AccountTokenSnapshot_OrderBy["TotalNetFlowRate"] = "totalNetFlowRate";
    AccountTokenSnapshot_OrderBy["TotalInflowRate"] = "totalInflowRate";
    AccountTokenSnapshot_OrderBy["TotalOutflowRate"] = "totalOutflowRate";
    AccountTokenSnapshot_OrderBy["TotalAmountStreamedInUntilUpdatedAt"] = "totalAmountStreamedInUntilUpdatedAt";
    AccountTokenSnapshot_OrderBy["TotalAmountStreamedOutUntilUpdatedAt"] = "totalAmountStreamedOutUntilUpdatedAt";
    AccountTokenSnapshot_OrderBy["TotalAmountStreamedUntilUpdatedAt"] = "totalAmountStreamedUntilUpdatedAt";
    AccountTokenSnapshot_OrderBy["TotalAmountTransferredUntilUpdatedAt"] = "totalAmountTransferredUntilUpdatedAt";
    AccountTokenSnapshot_OrderBy["Account"] = "account";
    AccountTokenSnapshot_OrderBy["Account__id"] = "account__id";
    AccountTokenSnapshot_OrderBy["Account__createdAtTimestamp"] = "account__createdAtTimestamp";
    AccountTokenSnapshot_OrderBy["Account__createdAtBlockNumber"] = "account__createdAtBlockNumber";
    AccountTokenSnapshot_OrderBy["Account__updatedAtTimestamp"] = "account__updatedAtTimestamp";
    AccountTokenSnapshot_OrderBy["Account__updatedAtBlockNumber"] = "account__updatedAtBlockNumber";
    AccountTokenSnapshot_OrderBy["Account__isSuperApp"] = "account__isSuperApp";
    AccountTokenSnapshot_OrderBy["Token"] = "token";
    AccountTokenSnapshot_OrderBy["Token__id"] = "token__id";
    AccountTokenSnapshot_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    AccountTokenSnapshot_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    AccountTokenSnapshot_OrderBy["Token__decimals"] = "token__decimals";
    AccountTokenSnapshot_OrderBy["Token__name"] = "token__name";
    AccountTokenSnapshot_OrderBy["Token__symbol"] = "token__symbol";
    AccountTokenSnapshot_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    AccountTokenSnapshot_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    AccountTokenSnapshot_OrderBy["Token__isListed"] = "token__isListed";
    AccountTokenSnapshot_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    AccountTokenSnapshot_OrderBy["FlowOperators"] = "flowOperators";
    AccountTokenSnapshot_OrderBy["AccountTokenSnapshotLogs"] = "accountTokenSnapshotLogs";
})(AccountTokenSnapshot_OrderBy = exports.AccountTokenSnapshot_OrderBy || (exports.AccountTokenSnapshot_OrderBy = {}));
var Account_OrderBy;
(function (Account_OrderBy) {
    Account_OrderBy["Id"] = "id";
    Account_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    Account_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    Account_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    Account_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    Account_OrderBy["IsSuperApp"] = "isSuperApp";
    Account_OrderBy["Inflows"] = "inflows";
    Account_OrderBy["Outflows"] = "outflows";
    Account_OrderBy["Subscriptions"] = "subscriptions";
    Account_OrderBy["PublishedIndexes"] = "publishedIndexes";
    Account_OrderBy["SentTransferEvents"] = "sentTransferEvents";
    Account_OrderBy["ReceivedTransferEvents"] = "receivedTransferEvents";
    Account_OrderBy["TokenUpgradedEvents"] = "tokenUpgradedEvents";
    Account_OrderBy["TokenDowngradedEvents"] = "tokenDowngradedEvents";
    Account_OrderBy["AccountTokenSnapshots"] = "accountTokenSnapshots";
})(Account_OrderBy = exports.Account_OrderBy || (exports.Account_OrderBy = {}));
var AgreementClassRegisteredEvent_OrderBy;
(function (AgreementClassRegisteredEvent_OrderBy) {
    AgreementClassRegisteredEvent_OrderBy["Id"] = "id";
    AgreementClassRegisteredEvent_OrderBy["TransactionHash"] = "transactionHash";
    AgreementClassRegisteredEvent_OrderBy["GasPrice"] = "gasPrice";
    AgreementClassRegisteredEvent_OrderBy["GasUsed"] = "gasUsed";
    AgreementClassRegisteredEvent_OrderBy["Timestamp"] = "timestamp";
    AgreementClassRegisteredEvent_OrderBy["Name"] = "name";
    AgreementClassRegisteredEvent_OrderBy["Addresses"] = "addresses";
    AgreementClassRegisteredEvent_OrderBy["BlockNumber"] = "blockNumber";
    AgreementClassRegisteredEvent_OrderBy["LogIndex"] = "logIndex";
    AgreementClassRegisteredEvent_OrderBy["Order"] = "order";
    AgreementClassRegisteredEvent_OrderBy["AgreementType"] = "agreementType";
    AgreementClassRegisteredEvent_OrderBy["Code"] = "code";
})(AgreementClassRegisteredEvent_OrderBy = exports.AgreementClassRegisteredEvent_OrderBy || (exports.AgreementClassRegisteredEvent_OrderBy = {}));
var AgreementClassUpdatedEvent_OrderBy;
(function (AgreementClassUpdatedEvent_OrderBy) {
    AgreementClassUpdatedEvent_OrderBy["Id"] = "id";
    AgreementClassUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    AgreementClassUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    AgreementClassUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    AgreementClassUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    AgreementClassUpdatedEvent_OrderBy["Name"] = "name";
    AgreementClassUpdatedEvent_OrderBy["Addresses"] = "addresses";
    AgreementClassUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    AgreementClassUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    AgreementClassUpdatedEvent_OrderBy["Order"] = "order";
    AgreementClassUpdatedEvent_OrderBy["AgreementType"] = "agreementType";
    AgreementClassUpdatedEvent_OrderBy["Code"] = "code";
})(AgreementClassUpdatedEvent_OrderBy = exports.AgreementClassUpdatedEvent_OrderBy || (exports.AgreementClassUpdatedEvent_OrderBy = {}));
var AgreementLiquidatedByEvent_OrderBy;
(function (AgreementLiquidatedByEvent_OrderBy) {
    AgreementLiquidatedByEvent_OrderBy["Id"] = "id";
    AgreementLiquidatedByEvent_OrderBy["TransactionHash"] = "transactionHash";
    AgreementLiquidatedByEvent_OrderBy["GasPrice"] = "gasPrice";
    AgreementLiquidatedByEvent_OrderBy["GasUsed"] = "gasUsed";
    AgreementLiquidatedByEvent_OrderBy["Timestamp"] = "timestamp";
    AgreementLiquidatedByEvent_OrderBy["Name"] = "name";
    AgreementLiquidatedByEvent_OrderBy["Addresses"] = "addresses";
    AgreementLiquidatedByEvent_OrderBy["BlockNumber"] = "blockNumber";
    AgreementLiquidatedByEvent_OrderBy["LogIndex"] = "logIndex";
    AgreementLiquidatedByEvent_OrderBy["Order"] = "order";
    AgreementLiquidatedByEvent_OrderBy["Token"] = "token";
    AgreementLiquidatedByEvent_OrderBy["LiquidatorAccount"] = "liquidatorAccount";
    AgreementLiquidatedByEvent_OrderBy["AgreementClass"] = "agreementClass";
    AgreementLiquidatedByEvent_OrderBy["AgreementId"] = "agreementId";
    AgreementLiquidatedByEvent_OrderBy["PenaltyAccount"] = "penaltyAccount";
    AgreementLiquidatedByEvent_OrderBy["BondAccount"] = "bondAccount";
    AgreementLiquidatedByEvent_OrderBy["RewardAmount"] = "rewardAmount";
    AgreementLiquidatedByEvent_OrderBy["BailoutAmount"] = "bailoutAmount";
    AgreementLiquidatedByEvent_OrderBy["Deposit"] = "deposit";
    AgreementLiquidatedByEvent_OrderBy["FlowRateAtLiquidation"] = "flowRateAtLiquidation";
})(AgreementLiquidatedByEvent_OrderBy = exports.AgreementLiquidatedByEvent_OrderBy || (exports.AgreementLiquidatedByEvent_OrderBy = {}));
var AgreementLiquidatedV2Event_OrderBy;
(function (AgreementLiquidatedV2Event_OrderBy) {
    AgreementLiquidatedV2Event_OrderBy["Id"] = "id";
    AgreementLiquidatedV2Event_OrderBy["TransactionHash"] = "transactionHash";
    AgreementLiquidatedV2Event_OrderBy["GasPrice"] = "gasPrice";
    AgreementLiquidatedV2Event_OrderBy["GasUsed"] = "gasUsed";
    AgreementLiquidatedV2Event_OrderBy["Timestamp"] = "timestamp";
    AgreementLiquidatedV2Event_OrderBy["Name"] = "name";
    AgreementLiquidatedV2Event_OrderBy["Addresses"] = "addresses";
    AgreementLiquidatedV2Event_OrderBy["BlockNumber"] = "blockNumber";
    AgreementLiquidatedV2Event_OrderBy["LogIndex"] = "logIndex";
    AgreementLiquidatedV2Event_OrderBy["Order"] = "order";
    AgreementLiquidatedV2Event_OrderBy["Token"] = "token";
    AgreementLiquidatedV2Event_OrderBy["AgreementClass"] = "agreementClass";
    AgreementLiquidatedV2Event_OrderBy["AgreementId"] = "agreementId";
    AgreementLiquidatedV2Event_OrderBy["LiquidatorAccount"] = "liquidatorAccount";
    AgreementLiquidatedV2Event_OrderBy["TargetAccount"] = "targetAccount";
    AgreementLiquidatedV2Event_OrderBy["RewardAmountReceiver"] = "rewardAmountReceiver";
    AgreementLiquidatedV2Event_OrderBy["RewardAmount"] = "rewardAmount";
    AgreementLiquidatedV2Event_OrderBy["TargetAccountBalanceDelta"] = "targetAccountBalanceDelta";
    AgreementLiquidatedV2Event_OrderBy["Version"] = "version";
    AgreementLiquidatedV2Event_OrderBy["LiquidationType"] = "liquidationType";
    AgreementLiquidatedV2Event_OrderBy["Deposit"] = "deposit";
    AgreementLiquidatedV2Event_OrderBy["FlowRateAtLiquidation"] = "flowRateAtLiquidation";
    AgreementLiquidatedV2Event_OrderBy["RewardAccount"] = "rewardAccount";
})(AgreementLiquidatedV2Event_OrderBy = exports.AgreementLiquidatedV2Event_OrderBy || (exports.AgreementLiquidatedV2Event_OrderBy = {}));
var AppRegisteredEvent_OrderBy;
(function (AppRegisteredEvent_OrderBy) {
    AppRegisteredEvent_OrderBy["Id"] = "id";
    AppRegisteredEvent_OrderBy["TransactionHash"] = "transactionHash";
    AppRegisteredEvent_OrderBy["GasPrice"] = "gasPrice";
    AppRegisteredEvent_OrderBy["GasUsed"] = "gasUsed";
    AppRegisteredEvent_OrderBy["Timestamp"] = "timestamp";
    AppRegisteredEvent_OrderBy["Name"] = "name";
    AppRegisteredEvent_OrderBy["Addresses"] = "addresses";
    AppRegisteredEvent_OrderBy["BlockNumber"] = "blockNumber";
    AppRegisteredEvent_OrderBy["LogIndex"] = "logIndex";
    AppRegisteredEvent_OrderBy["Order"] = "order";
    AppRegisteredEvent_OrderBy["App"] = "app";
})(AppRegisteredEvent_OrderBy = exports.AppRegisteredEvent_OrderBy || (exports.AppRegisteredEvent_OrderBy = {}));
var BondIncreasedEvent_OrderBy;
(function (BondIncreasedEvent_OrderBy) {
    BondIncreasedEvent_OrderBy["Id"] = "id";
    BondIncreasedEvent_OrderBy["TransactionHash"] = "transactionHash";
    BondIncreasedEvent_OrderBy["GasPrice"] = "gasPrice";
    BondIncreasedEvent_OrderBy["GasUsed"] = "gasUsed";
    BondIncreasedEvent_OrderBy["Timestamp"] = "timestamp";
    BondIncreasedEvent_OrderBy["Name"] = "name";
    BondIncreasedEvent_OrderBy["Addresses"] = "addresses";
    BondIncreasedEvent_OrderBy["BlockNumber"] = "blockNumber";
    BondIncreasedEvent_OrderBy["LogIndex"] = "logIndex";
    BondIncreasedEvent_OrderBy["Order"] = "order";
    BondIncreasedEvent_OrderBy["Token"] = "token";
    BondIncreasedEvent_OrderBy["AdditionalBond"] = "additionalBond";
})(BondIncreasedEvent_OrderBy = exports.BondIncreasedEvent_OrderBy || (exports.BondIncreasedEvent_OrderBy = {}));
var BurnedEvent_OrderBy;
(function (BurnedEvent_OrderBy) {
    BurnedEvent_OrderBy["Id"] = "id";
    BurnedEvent_OrderBy["TransactionHash"] = "transactionHash";
    BurnedEvent_OrderBy["GasPrice"] = "gasPrice";
    BurnedEvent_OrderBy["GasUsed"] = "gasUsed";
    BurnedEvent_OrderBy["Timestamp"] = "timestamp";
    BurnedEvent_OrderBy["Name"] = "name";
    BurnedEvent_OrderBy["Addresses"] = "addresses";
    BurnedEvent_OrderBy["BlockNumber"] = "blockNumber";
    BurnedEvent_OrderBy["LogIndex"] = "logIndex";
    BurnedEvent_OrderBy["Order"] = "order";
    BurnedEvent_OrderBy["Operator"] = "operator";
    BurnedEvent_OrderBy["From"] = "from";
    BurnedEvent_OrderBy["Token"] = "token";
    BurnedEvent_OrderBy["Amount"] = "amount";
    BurnedEvent_OrderBy["Data"] = "data";
    BurnedEvent_OrderBy["OperatorData"] = "operatorData";
})(BurnedEvent_OrderBy = exports.BurnedEvent_OrderBy || (exports.BurnedEvent_OrderBy = {}));
var CfAv1LiquidationPeriodChangedEvent_OrderBy;
(function (CfAv1LiquidationPeriodChangedEvent_OrderBy) {
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Id"] = "id";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Timestamp"] = "timestamp";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Name"] = "name";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Addresses"] = "addresses";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Order"] = "order";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["LogIndex"] = "logIndex";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["Host"] = "host";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["SuperToken"] = "superToken";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    CfAv1LiquidationPeriodChangedEvent_OrderBy["LiquidationPeriod"] = "liquidationPeriod";
})(CfAv1LiquidationPeriodChangedEvent_OrderBy = exports.CfAv1LiquidationPeriodChangedEvent_OrderBy || (exports.CfAv1LiquidationPeriodChangedEvent_OrderBy = {}));
var ConfigChangedEvent_OrderBy;
(function (ConfigChangedEvent_OrderBy) {
    ConfigChangedEvent_OrderBy["Id"] = "id";
    ConfigChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    ConfigChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    ConfigChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    ConfigChangedEvent_OrderBy["Timestamp"] = "timestamp";
    ConfigChangedEvent_OrderBy["Name"] = "name";
    ConfigChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    ConfigChangedEvent_OrderBy["Addresses"] = "addresses";
    ConfigChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    ConfigChangedEvent_OrderBy["LogIndex"] = "logIndex";
    ConfigChangedEvent_OrderBy["Order"] = "order";
    ConfigChangedEvent_OrderBy["Host"] = "host";
    ConfigChangedEvent_OrderBy["SuperToken"] = "superToken";
    ConfigChangedEvent_OrderBy["Key"] = "key";
    ConfigChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    ConfigChangedEvent_OrderBy["Value"] = "value";
})(ConfigChangedEvent_OrderBy = exports.ConfigChangedEvent_OrderBy || (exports.ConfigChangedEvent_OrderBy = {}));
var CustomSuperTokenCreatedEvent_OrderBy;
(function (CustomSuperTokenCreatedEvent_OrderBy) {
    CustomSuperTokenCreatedEvent_OrderBy["Id"] = "id";
    CustomSuperTokenCreatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    CustomSuperTokenCreatedEvent_OrderBy["GasPrice"] = "gasPrice";
    CustomSuperTokenCreatedEvent_OrderBy["GasUsed"] = "gasUsed";
    CustomSuperTokenCreatedEvent_OrderBy["Timestamp"] = "timestamp";
    CustomSuperTokenCreatedEvent_OrderBy["Name"] = "name";
    CustomSuperTokenCreatedEvent_OrderBy["Addresses"] = "addresses";
    CustomSuperTokenCreatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    CustomSuperTokenCreatedEvent_OrderBy["LogIndex"] = "logIndex";
    CustomSuperTokenCreatedEvent_OrderBy["Order"] = "order";
    CustomSuperTokenCreatedEvent_OrderBy["Token"] = "token";
})(CustomSuperTokenCreatedEvent_OrderBy = exports.CustomSuperTokenCreatedEvent_OrderBy || (exports.CustomSuperTokenCreatedEvent_OrderBy = {}));
var Event_OrderBy;
(function (Event_OrderBy) {
    Event_OrderBy["Id"] = "id";
    Event_OrderBy["BlockNumber"] = "blockNumber";
    Event_OrderBy["LogIndex"] = "logIndex";
    Event_OrderBy["Order"] = "order";
    Event_OrderBy["Name"] = "name";
    Event_OrderBy["Addresses"] = "addresses";
    Event_OrderBy["Timestamp"] = "timestamp";
    Event_OrderBy["TransactionHash"] = "transactionHash";
    Event_OrderBy["GasPrice"] = "gasPrice";
    Event_OrderBy["GasUsed"] = "gasUsed";
})(Event_OrderBy = exports.Event_OrderBy || (exports.Event_OrderBy = {}));
var ExitRateChangedEvent_OrderBy;
(function (ExitRateChangedEvent_OrderBy) {
    ExitRateChangedEvent_OrderBy["Id"] = "id";
    ExitRateChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    ExitRateChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    ExitRateChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    ExitRateChangedEvent_OrderBy["Timestamp"] = "timestamp";
    ExitRateChangedEvent_OrderBy["Name"] = "name";
    ExitRateChangedEvent_OrderBy["Addresses"] = "addresses";
    ExitRateChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    ExitRateChangedEvent_OrderBy["LogIndex"] = "logIndex";
    ExitRateChangedEvent_OrderBy["Order"] = "order";
    ExitRateChangedEvent_OrderBy["Token"] = "token";
    ExitRateChangedEvent_OrderBy["ExitRate"] = "exitRate";
})(ExitRateChangedEvent_OrderBy = exports.ExitRateChangedEvent_OrderBy || (exports.ExitRateChangedEvent_OrderBy = {}));
var FlowOperatorUpdatedEvent_OrderBy;
(function (FlowOperatorUpdatedEvent_OrderBy) {
    FlowOperatorUpdatedEvent_OrderBy["Id"] = "id";
    FlowOperatorUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    FlowOperatorUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    FlowOperatorUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    FlowOperatorUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    FlowOperatorUpdatedEvent_OrderBy["Name"] = "name";
    FlowOperatorUpdatedEvent_OrderBy["Addresses"] = "addresses";
    FlowOperatorUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    FlowOperatorUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    FlowOperatorUpdatedEvent_OrderBy["Order"] = "order";
    FlowOperatorUpdatedEvent_OrderBy["Token"] = "token";
    FlowOperatorUpdatedEvent_OrderBy["Sender"] = "sender";
    FlowOperatorUpdatedEvent_OrderBy["Permissions"] = "permissions";
    FlowOperatorUpdatedEvent_OrderBy["FlowRateAllowance"] = "flowRateAllowance";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator"] = "flowOperator";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__id"] = "flowOperator__id";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__createdAtTimestamp"] = "flowOperator__createdAtTimestamp";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__createdAtBlockNumber"] = "flowOperator__createdAtBlockNumber";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__updatedAtTimestamp"] = "flowOperator__updatedAtTimestamp";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__updatedAtBlockNumber"] = "flowOperator__updatedAtBlockNumber";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__permissions"] = "flowOperator__permissions";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__flowRateAllowanceGranted"] = "flowOperator__flowRateAllowanceGranted";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__flowRateAllowanceRemaining"] = "flowOperator__flowRateAllowanceRemaining";
    FlowOperatorUpdatedEvent_OrderBy["FlowOperator__flowOperator"] = "flowOperator__flowOperator";
})(FlowOperatorUpdatedEvent_OrderBy = exports.FlowOperatorUpdatedEvent_OrderBy || (exports.FlowOperatorUpdatedEvent_OrderBy = {}));
var FlowOperator_OrderBy;
(function (FlowOperator_OrderBy) {
    FlowOperator_OrderBy["Id"] = "id";
    FlowOperator_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    FlowOperator_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    FlowOperator_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    FlowOperator_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    FlowOperator_OrderBy["Permissions"] = "permissions";
    FlowOperator_OrderBy["FlowRateAllowanceGranted"] = "flowRateAllowanceGranted";
    FlowOperator_OrderBy["FlowRateAllowanceRemaining"] = "flowRateAllowanceRemaining";
    FlowOperator_OrderBy["FlowOperator"] = "flowOperator";
    FlowOperator_OrderBy["Sender"] = "sender";
    FlowOperator_OrderBy["Sender__id"] = "sender__id";
    FlowOperator_OrderBy["Sender__createdAtTimestamp"] = "sender__createdAtTimestamp";
    FlowOperator_OrderBy["Sender__createdAtBlockNumber"] = "sender__createdAtBlockNumber";
    FlowOperator_OrderBy["Sender__updatedAtTimestamp"] = "sender__updatedAtTimestamp";
    FlowOperator_OrderBy["Sender__updatedAtBlockNumber"] = "sender__updatedAtBlockNumber";
    FlowOperator_OrderBy["Sender__isSuperApp"] = "sender__isSuperApp";
    FlowOperator_OrderBy["Token"] = "token";
    FlowOperator_OrderBy["Token__id"] = "token__id";
    FlowOperator_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    FlowOperator_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    FlowOperator_OrderBy["Token__decimals"] = "token__decimals";
    FlowOperator_OrderBy["Token__name"] = "token__name";
    FlowOperator_OrderBy["Token__symbol"] = "token__symbol";
    FlowOperator_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    FlowOperator_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    FlowOperator_OrderBy["Token__isListed"] = "token__isListed";
    FlowOperator_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    FlowOperator_OrderBy["AccountTokenSnapshot"] = "accountTokenSnapshot";
    FlowOperator_OrderBy["AccountTokenSnapshot__id"] = "accountTokenSnapshot__id";
    FlowOperator_OrderBy["AccountTokenSnapshot__updatedAtTimestamp"] = "accountTokenSnapshot__updatedAtTimestamp";
    FlowOperator_OrderBy["AccountTokenSnapshot__updatedAtBlockNumber"] = "accountTokenSnapshot__updatedAtBlockNumber";
    FlowOperator_OrderBy["AccountTokenSnapshot__isLiquidationEstimateOptimistic"] = "accountTokenSnapshot__isLiquidationEstimateOptimistic";
    FlowOperator_OrderBy["AccountTokenSnapshot__maybeCriticalAtTimestamp"] = "accountTokenSnapshot__maybeCriticalAtTimestamp";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalNumberOfActiveStreams"] = "accountTokenSnapshot__totalNumberOfActiveStreams";
    FlowOperator_OrderBy["AccountTokenSnapshot__activeOutgoingStreamCount"] = "accountTokenSnapshot__activeOutgoingStreamCount";
    FlowOperator_OrderBy["AccountTokenSnapshot__activeIncomingStreamCount"] = "accountTokenSnapshot__activeIncomingStreamCount";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalNumberOfClosedStreams"] = "accountTokenSnapshot__totalNumberOfClosedStreams";
    FlowOperator_OrderBy["AccountTokenSnapshot__inactiveOutgoingStreamCount"] = "accountTokenSnapshot__inactiveOutgoingStreamCount";
    FlowOperator_OrderBy["AccountTokenSnapshot__inactiveIncomingStreamCount"] = "accountTokenSnapshot__inactiveIncomingStreamCount";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalSubscriptionsWithUnits"] = "accountTokenSnapshot__totalSubscriptionsWithUnits";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalApprovedSubscriptions"] = "accountTokenSnapshot__totalApprovedSubscriptions";
    FlowOperator_OrderBy["AccountTokenSnapshot__balanceUntilUpdatedAt"] = "accountTokenSnapshot__balanceUntilUpdatedAt";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalDeposit"] = "accountTokenSnapshot__totalDeposit";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalNetFlowRate"] = "accountTokenSnapshot__totalNetFlowRate";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalInflowRate"] = "accountTokenSnapshot__totalInflowRate";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalOutflowRate"] = "accountTokenSnapshot__totalOutflowRate";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalAmountStreamedInUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedInUntilUpdatedAt";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalAmountStreamedOutUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedOutUntilUpdatedAt";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalAmountStreamedUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountStreamedUntilUpdatedAt";
    FlowOperator_OrderBy["AccountTokenSnapshot__totalAmountTransferredUntilUpdatedAt"] = "accountTokenSnapshot__totalAmountTransferredUntilUpdatedAt";
    FlowOperator_OrderBy["FlowOperatorUpdatedEvents"] = "flowOperatorUpdatedEvents";
})(FlowOperator_OrderBy = exports.FlowOperator_OrderBy || (exports.FlowOperator_OrderBy = {}));
var FlowUpdatedEvent_OrderBy;
(function (FlowUpdatedEvent_OrderBy) {
    FlowUpdatedEvent_OrderBy["Id"] = "id";
    FlowUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    FlowUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    FlowUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    FlowUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    FlowUpdatedEvent_OrderBy["Name"] = "name";
    FlowUpdatedEvent_OrderBy["Addresses"] = "addresses";
    FlowUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    FlowUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    FlowUpdatedEvent_OrderBy["Order"] = "order";
    FlowUpdatedEvent_OrderBy["Token"] = "token";
    FlowUpdatedEvent_OrderBy["Sender"] = "sender";
    FlowUpdatedEvent_OrderBy["Receiver"] = "receiver";
    FlowUpdatedEvent_OrderBy["FlowOperator"] = "flowOperator";
    FlowUpdatedEvent_OrderBy["FlowRate"] = "flowRate";
    FlowUpdatedEvent_OrderBy["TotalSenderFlowRate"] = "totalSenderFlowRate";
    FlowUpdatedEvent_OrderBy["TotalReceiverFlowRate"] = "totalReceiverFlowRate";
    FlowUpdatedEvent_OrderBy["Deposit"] = "deposit";
    FlowUpdatedEvent_OrderBy["UserData"] = "userData";
    FlowUpdatedEvent_OrderBy["OldFlowRate"] = "oldFlowRate";
    FlowUpdatedEvent_OrderBy["Type"] = "type";
    FlowUpdatedEvent_OrderBy["TotalAmountStreamedUntilTimestamp"] = "totalAmountStreamedUntilTimestamp";
    FlowUpdatedEvent_OrderBy["Stream"] = "stream";
    FlowUpdatedEvent_OrderBy["Stream__id"] = "stream__id";
    FlowUpdatedEvent_OrderBy["Stream__createdAtTimestamp"] = "stream__createdAtTimestamp";
    FlowUpdatedEvent_OrderBy["Stream__createdAtBlockNumber"] = "stream__createdAtBlockNumber";
    FlowUpdatedEvent_OrderBy["Stream__updatedAtTimestamp"] = "stream__updatedAtTimestamp";
    FlowUpdatedEvent_OrderBy["Stream__updatedAtBlockNumber"] = "stream__updatedAtBlockNumber";
    FlowUpdatedEvent_OrderBy["Stream__currentFlowRate"] = "stream__currentFlowRate";
    FlowUpdatedEvent_OrderBy["Stream__deposit"] = "stream__deposit";
    FlowUpdatedEvent_OrderBy["Stream__streamedUntilUpdatedAt"] = "stream__streamedUntilUpdatedAt";
    FlowUpdatedEvent_OrderBy["Stream__userData"] = "stream__userData";
})(FlowUpdatedEvent_OrderBy = exports.FlowUpdatedEvent_OrderBy || (exports.FlowUpdatedEvent_OrderBy = {}));
var GovernanceReplacedEvent_OrderBy;
(function (GovernanceReplacedEvent_OrderBy) {
    GovernanceReplacedEvent_OrderBy["Id"] = "id";
    GovernanceReplacedEvent_OrderBy["TransactionHash"] = "transactionHash";
    GovernanceReplacedEvent_OrderBy["GasPrice"] = "gasPrice";
    GovernanceReplacedEvent_OrderBy["GasUsed"] = "gasUsed";
    GovernanceReplacedEvent_OrderBy["Timestamp"] = "timestamp";
    GovernanceReplacedEvent_OrderBy["Name"] = "name";
    GovernanceReplacedEvent_OrderBy["BlockNumber"] = "blockNumber";
    GovernanceReplacedEvent_OrderBy["LogIndex"] = "logIndex";
    GovernanceReplacedEvent_OrderBy["Order"] = "order";
    GovernanceReplacedEvent_OrderBy["Addresses"] = "addresses";
    GovernanceReplacedEvent_OrderBy["OldGovernance"] = "oldGovernance";
    GovernanceReplacedEvent_OrderBy["NewGovernance"] = "newGovernance";
})(GovernanceReplacedEvent_OrderBy = exports.GovernanceReplacedEvent_OrderBy || (exports.GovernanceReplacedEvent_OrderBy = {}));
var IndexCreatedEvent_OrderBy;
(function (IndexCreatedEvent_OrderBy) {
    IndexCreatedEvent_OrderBy["Id"] = "id";
    IndexCreatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexCreatedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexCreatedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexCreatedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexCreatedEvent_OrderBy["Name"] = "name";
    IndexCreatedEvent_OrderBy["Addresses"] = "addresses";
    IndexCreatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexCreatedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexCreatedEvent_OrderBy["Order"] = "order";
    IndexCreatedEvent_OrderBy["Token"] = "token";
    IndexCreatedEvent_OrderBy["Publisher"] = "publisher";
    IndexCreatedEvent_OrderBy["IndexId"] = "indexId";
    IndexCreatedEvent_OrderBy["UserData"] = "userData";
    IndexCreatedEvent_OrderBy["Index"] = "index";
    IndexCreatedEvent_OrderBy["Index__id"] = "index__id";
    IndexCreatedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexCreatedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexCreatedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexCreatedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexCreatedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexCreatedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexCreatedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexCreatedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexCreatedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexCreatedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexCreatedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexCreatedEvent_OrderBy = exports.IndexCreatedEvent_OrderBy || (exports.IndexCreatedEvent_OrderBy = {}));
var IndexDistributionClaimedEvent_OrderBy;
(function (IndexDistributionClaimedEvent_OrderBy) {
    IndexDistributionClaimedEvent_OrderBy["Id"] = "id";
    IndexDistributionClaimedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexDistributionClaimedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexDistributionClaimedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexDistributionClaimedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexDistributionClaimedEvent_OrderBy["Name"] = "name";
    IndexDistributionClaimedEvent_OrderBy["Addresses"] = "addresses";
    IndexDistributionClaimedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexDistributionClaimedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexDistributionClaimedEvent_OrderBy["Order"] = "order";
    IndexDistributionClaimedEvent_OrderBy["Token"] = "token";
    IndexDistributionClaimedEvent_OrderBy["Publisher"] = "publisher";
    IndexDistributionClaimedEvent_OrderBy["IndexId"] = "indexId";
    IndexDistributionClaimedEvent_OrderBy["Subscriber"] = "subscriber";
    IndexDistributionClaimedEvent_OrderBy["Amount"] = "amount";
    IndexDistributionClaimedEvent_OrderBy["Index"] = "index";
    IndexDistributionClaimedEvent_OrderBy["Index__id"] = "index__id";
    IndexDistributionClaimedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexDistributionClaimedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexDistributionClaimedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexDistributionClaimedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexDistributionClaimedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexDistributionClaimedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexDistributionClaimedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexDistributionClaimedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexDistributionClaimedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexDistributionClaimedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexDistributionClaimedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexDistributionClaimedEvent_OrderBy = exports.IndexDistributionClaimedEvent_OrderBy || (exports.IndexDistributionClaimedEvent_OrderBy = {}));
var IndexSubscribedEvent_OrderBy;
(function (IndexSubscribedEvent_OrderBy) {
    IndexSubscribedEvent_OrderBy["Id"] = "id";
    IndexSubscribedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexSubscribedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexSubscribedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexSubscribedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexSubscribedEvent_OrderBy["Name"] = "name";
    IndexSubscribedEvent_OrderBy["Addresses"] = "addresses";
    IndexSubscribedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexSubscribedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexSubscribedEvent_OrderBy["Order"] = "order";
    IndexSubscribedEvent_OrderBy["Token"] = "token";
    IndexSubscribedEvent_OrderBy["Publisher"] = "publisher";
    IndexSubscribedEvent_OrderBy["IndexId"] = "indexId";
    IndexSubscribedEvent_OrderBy["Subscriber"] = "subscriber";
    IndexSubscribedEvent_OrderBy["UserData"] = "userData";
    IndexSubscribedEvent_OrderBy["Index"] = "index";
    IndexSubscribedEvent_OrderBy["Index__id"] = "index__id";
    IndexSubscribedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexSubscribedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexSubscribedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexSubscribedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexSubscribedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexSubscribedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexSubscribedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexSubscribedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexSubscribedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexSubscribedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexSubscribedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexSubscribedEvent_OrderBy = exports.IndexSubscribedEvent_OrderBy || (exports.IndexSubscribedEvent_OrderBy = {}));
var IndexSubscription_OrderBy;
(function (IndexSubscription_OrderBy) {
    IndexSubscription_OrderBy["Id"] = "id";
    IndexSubscription_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    IndexSubscription_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    IndexSubscription_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    IndexSubscription_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    IndexSubscription_OrderBy["Subscriber"] = "subscriber";
    IndexSubscription_OrderBy["Subscriber__id"] = "subscriber__id";
    IndexSubscription_OrderBy["Subscriber__createdAtTimestamp"] = "subscriber__createdAtTimestamp";
    IndexSubscription_OrderBy["Subscriber__createdAtBlockNumber"] = "subscriber__createdAtBlockNumber";
    IndexSubscription_OrderBy["Subscriber__updatedAtTimestamp"] = "subscriber__updatedAtTimestamp";
    IndexSubscription_OrderBy["Subscriber__updatedAtBlockNumber"] = "subscriber__updatedAtBlockNumber";
    IndexSubscription_OrderBy["Subscriber__isSuperApp"] = "subscriber__isSuperApp";
    IndexSubscription_OrderBy["Approved"] = "approved";
    IndexSubscription_OrderBy["Units"] = "units";
    IndexSubscription_OrderBy["TotalAmountReceivedUntilUpdatedAt"] = "totalAmountReceivedUntilUpdatedAt";
    IndexSubscription_OrderBy["IndexValueUntilUpdatedAt"] = "indexValueUntilUpdatedAt";
    IndexSubscription_OrderBy["Index"] = "index";
    IndexSubscription_OrderBy["Index__id"] = "index__id";
    IndexSubscription_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexSubscription_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexSubscription_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexSubscription_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexSubscription_OrderBy["Index__indexId"] = "index__indexId";
    IndexSubscription_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexSubscription_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexSubscription_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexSubscription_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexSubscription_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexSubscription_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
    IndexSubscription_OrderBy["SubscriptionApprovedEvents"] = "subscriptionApprovedEvents";
    IndexSubscription_OrderBy["SubscriptionDistributionClaimedEvents"] = "subscriptionDistributionClaimedEvents";
    IndexSubscription_OrderBy["SubscriptionRevokedEvents"] = "subscriptionRevokedEvents";
    IndexSubscription_OrderBy["SubscriptionUnitsUpdatedEvents"] = "subscriptionUnitsUpdatedEvents";
})(IndexSubscription_OrderBy = exports.IndexSubscription_OrderBy || (exports.IndexSubscription_OrderBy = {}));
var IndexUnitsUpdatedEvent_OrderBy;
(function (IndexUnitsUpdatedEvent_OrderBy) {
    IndexUnitsUpdatedEvent_OrderBy["Id"] = "id";
    IndexUnitsUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexUnitsUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexUnitsUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexUnitsUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexUnitsUpdatedEvent_OrderBy["Name"] = "name";
    IndexUnitsUpdatedEvent_OrderBy["Addresses"] = "addresses";
    IndexUnitsUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexUnitsUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexUnitsUpdatedEvent_OrderBy["Order"] = "order";
    IndexUnitsUpdatedEvent_OrderBy["Token"] = "token";
    IndexUnitsUpdatedEvent_OrderBy["Publisher"] = "publisher";
    IndexUnitsUpdatedEvent_OrderBy["IndexId"] = "indexId";
    IndexUnitsUpdatedEvent_OrderBy["Subscriber"] = "subscriber";
    IndexUnitsUpdatedEvent_OrderBy["Units"] = "units";
    IndexUnitsUpdatedEvent_OrderBy["UserData"] = "userData";
    IndexUnitsUpdatedEvent_OrderBy["OldUnits"] = "oldUnits";
    IndexUnitsUpdatedEvent_OrderBy["Index"] = "index";
    IndexUnitsUpdatedEvent_OrderBy["Index__id"] = "index__id";
    IndexUnitsUpdatedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexUnitsUpdatedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexUnitsUpdatedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexUnitsUpdatedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexUnitsUpdatedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexUnitsUpdatedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexUnitsUpdatedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexUnitsUpdatedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexUnitsUpdatedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexUnitsUpdatedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexUnitsUpdatedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexUnitsUpdatedEvent_OrderBy = exports.IndexUnitsUpdatedEvent_OrderBy || (exports.IndexUnitsUpdatedEvent_OrderBy = {}));
var IndexUnsubscribedEvent_OrderBy;
(function (IndexUnsubscribedEvent_OrderBy) {
    IndexUnsubscribedEvent_OrderBy["Id"] = "id";
    IndexUnsubscribedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexUnsubscribedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexUnsubscribedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexUnsubscribedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexUnsubscribedEvent_OrderBy["Name"] = "name";
    IndexUnsubscribedEvent_OrderBy["Addresses"] = "addresses";
    IndexUnsubscribedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexUnsubscribedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexUnsubscribedEvent_OrderBy["Order"] = "order";
    IndexUnsubscribedEvent_OrderBy["Token"] = "token";
    IndexUnsubscribedEvent_OrderBy["Publisher"] = "publisher";
    IndexUnsubscribedEvent_OrderBy["IndexId"] = "indexId";
    IndexUnsubscribedEvent_OrderBy["Subscriber"] = "subscriber";
    IndexUnsubscribedEvent_OrderBy["UserData"] = "userData";
    IndexUnsubscribedEvent_OrderBy["Index"] = "index";
    IndexUnsubscribedEvent_OrderBy["Index__id"] = "index__id";
    IndexUnsubscribedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexUnsubscribedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexUnsubscribedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexUnsubscribedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexUnsubscribedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexUnsubscribedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexUnsubscribedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexUnsubscribedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexUnsubscribedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexUnsubscribedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexUnsubscribedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexUnsubscribedEvent_OrderBy = exports.IndexUnsubscribedEvent_OrderBy || (exports.IndexUnsubscribedEvent_OrderBy = {}));
var IndexUpdatedEvent_OrderBy;
(function (IndexUpdatedEvent_OrderBy) {
    IndexUpdatedEvent_OrderBy["Id"] = "id";
    IndexUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    IndexUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    IndexUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    IndexUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    IndexUpdatedEvent_OrderBy["Name"] = "name";
    IndexUpdatedEvent_OrderBy["Addresses"] = "addresses";
    IndexUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    IndexUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    IndexUpdatedEvent_OrderBy["Order"] = "order";
    IndexUpdatedEvent_OrderBy["Token"] = "token";
    IndexUpdatedEvent_OrderBy["Publisher"] = "publisher";
    IndexUpdatedEvent_OrderBy["IndexId"] = "indexId";
    IndexUpdatedEvent_OrderBy["OldIndexValue"] = "oldIndexValue";
    IndexUpdatedEvent_OrderBy["NewIndexValue"] = "newIndexValue";
    IndexUpdatedEvent_OrderBy["TotalUnitsPending"] = "totalUnitsPending";
    IndexUpdatedEvent_OrderBy["TotalUnitsApproved"] = "totalUnitsApproved";
    IndexUpdatedEvent_OrderBy["UserData"] = "userData";
    IndexUpdatedEvent_OrderBy["Index"] = "index";
    IndexUpdatedEvent_OrderBy["Index__id"] = "index__id";
    IndexUpdatedEvent_OrderBy["Index__createdAtTimestamp"] = "index__createdAtTimestamp";
    IndexUpdatedEvent_OrderBy["Index__createdAtBlockNumber"] = "index__createdAtBlockNumber";
    IndexUpdatedEvent_OrderBy["Index__updatedAtTimestamp"] = "index__updatedAtTimestamp";
    IndexUpdatedEvent_OrderBy["Index__updatedAtBlockNumber"] = "index__updatedAtBlockNumber";
    IndexUpdatedEvent_OrderBy["Index__indexId"] = "index__indexId";
    IndexUpdatedEvent_OrderBy["Index__indexValue"] = "index__indexValue";
    IndexUpdatedEvent_OrderBy["Index__totalSubscriptionsWithUnits"] = "index__totalSubscriptionsWithUnits";
    IndexUpdatedEvent_OrderBy["Index__totalUnitsPending"] = "index__totalUnitsPending";
    IndexUpdatedEvent_OrderBy["Index__totalUnitsApproved"] = "index__totalUnitsApproved";
    IndexUpdatedEvent_OrderBy["Index__totalUnits"] = "index__totalUnits";
    IndexUpdatedEvent_OrderBy["Index__totalAmountDistributedUntilUpdatedAt"] = "index__totalAmountDistributedUntilUpdatedAt";
})(IndexUpdatedEvent_OrderBy = exports.IndexUpdatedEvent_OrderBy || (exports.IndexUpdatedEvent_OrderBy = {}));
var Index_OrderBy;
(function (Index_OrderBy) {
    Index_OrderBy["Id"] = "id";
    Index_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    Index_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    Index_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    Index_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    Index_OrderBy["IndexId"] = "indexId";
    Index_OrderBy["IndexValue"] = "indexValue";
    Index_OrderBy["TotalSubscriptionsWithUnits"] = "totalSubscriptionsWithUnits";
    Index_OrderBy["TotalUnitsPending"] = "totalUnitsPending";
    Index_OrderBy["TotalUnitsApproved"] = "totalUnitsApproved";
    Index_OrderBy["TotalUnits"] = "totalUnits";
    Index_OrderBy["TotalAmountDistributedUntilUpdatedAt"] = "totalAmountDistributedUntilUpdatedAt";
    Index_OrderBy["Token"] = "token";
    Index_OrderBy["Token__id"] = "token__id";
    Index_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    Index_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    Index_OrderBy["Token__decimals"] = "token__decimals";
    Index_OrderBy["Token__name"] = "token__name";
    Index_OrderBy["Token__symbol"] = "token__symbol";
    Index_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    Index_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    Index_OrderBy["Token__isListed"] = "token__isListed";
    Index_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    Index_OrderBy["Publisher"] = "publisher";
    Index_OrderBy["Publisher__id"] = "publisher__id";
    Index_OrderBy["Publisher__createdAtTimestamp"] = "publisher__createdAtTimestamp";
    Index_OrderBy["Publisher__createdAtBlockNumber"] = "publisher__createdAtBlockNumber";
    Index_OrderBy["Publisher__updatedAtTimestamp"] = "publisher__updatedAtTimestamp";
    Index_OrderBy["Publisher__updatedAtBlockNumber"] = "publisher__updatedAtBlockNumber";
    Index_OrderBy["Publisher__isSuperApp"] = "publisher__isSuperApp";
    Index_OrderBy["Subscriptions"] = "subscriptions";
    Index_OrderBy["IndexCreatedEvent"] = "indexCreatedEvent";
    Index_OrderBy["IndexCreatedEvent__id"] = "indexCreatedEvent__id";
    Index_OrderBy["IndexCreatedEvent__transactionHash"] = "indexCreatedEvent__transactionHash";
    Index_OrderBy["IndexCreatedEvent__gasPrice"] = "indexCreatedEvent__gasPrice";
    Index_OrderBy["IndexCreatedEvent__gasUsed"] = "indexCreatedEvent__gasUsed";
    Index_OrderBy["IndexCreatedEvent__timestamp"] = "indexCreatedEvent__timestamp";
    Index_OrderBy["IndexCreatedEvent__name"] = "indexCreatedEvent__name";
    Index_OrderBy["IndexCreatedEvent__blockNumber"] = "indexCreatedEvent__blockNumber";
    Index_OrderBy["IndexCreatedEvent__logIndex"] = "indexCreatedEvent__logIndex";
    Index_OrderBy["IndexCreatedEvent__order"] = "indexCreatedEvent__order";
    Index_OrderBy["IndexCreatedEvent__token"] = "indexCreatedEvent__token";
    Index_OrderBy["IndexCreatedEvent__publisher"] = "indexCreatedEvent__publisher";
    Index_OrderBy["IndexCreatedEvent__indexId"] = "indexCreatedEvent__indexId";
    Index_OrderBy["IndexCreatedEvent__userData"] = "indexCreatedEvent__userData";
    Index_OrderBy["IndexDistributionClaimedEvents"] = "indexDistributionClaimedEvents";
    Index_OrderBy["IndexUpdatedEvents"] = "indexUpdatedEvents";
    Index_OrderBy["IndexSubscribedEvents"] = "indexSubscribedEvents";
    Index_OrderBy["IndexUnitsUpdatedEvents"] = "indexUnitsUpdatedEvents";
    Index_OrderBy["IndexUnsubscribedEvents"] = "indexUnsubscribedEvents";
})(Index_OrderBy = exports.Index_OrderBy || (exports.Index_OrderBy = {}));
var JailEvent_OrderBy;
(function (JailEvent_OrderBy) {
    JailEvent_OrderBy["Id"] = "id";
    JailEvent_OrderBy["TransactionHash"] = "transactionHash";
    JailEvent_OrderBy["GasPrice"] = "gasPrice";
    JailEvent_OrderBy["GasUsed"] = "gasUsed";
    JailEvent_OrderBy["Timestamp"] = "timestamp";
    JailEvent_OrderBy["Name"] = "name";
    JailEvent_OrderBy["Addresses"] = "addresses";
    JailEvent_OrderBy["BlockNumber"] = "blockNumber";
    JailEvent_OrderBy["LogIndex"] = "logIndex";
    JailEvent_OrderBy["Order"] = "order";
    JailEvent_OrderBy["App"] = "app";
    JailEvent_OrderBy["Reason"] = "reason";
})(JailEvent_OrderBy = exports.JailEvent_OrderBy || (exports.JailEvent_OrderBy = {}));
var MintedEvent_OrderBy;
(function (MintedEvent_OrderBy) {
    MintedEvent_OrderBy["Id"] = "id";
    MintedEvent_OrderBy["TransactionHash"] = "transactionHash";
    MintedEvent_OrderBy["GasPrice"] = "gasPrice";
    MintedEvent_OrderBy["GasUsed"] = "gasUsed";
    MintedEvent_OrderBy["Timestamp"] = "timestamp";
    MintedEvent_OrderBy["Name"] = "name";
    MintedEvent_OrderBy["Addresses"] = "addresses";
    MintedEvent_OrderBy["BlockNumber"] = "blockNumber";
    MintedEvent_OrderBy["LogIndex"] = "logIndex";
    MintedEvent_OrderBy["Order"] = "order";
    MintedEvent_OrderBy["Operator"] = "operator";
    MintedEvent_OrderBy["To"] = "to";
    MintedEvent_OrderBy["Amount"] = "amount";
    MintedEvent_OrderBy["Data"] = "data";
    MintedEvent_OrderBy["Token"] = "token";
    MintedEvent_OrderBy["OperatorData"] = "operatorData";
})(MintedEvent_OrderBy = exports.MintedEvent_OrderBy || (exports.MintedEvent_OrderBy = {}));
var NewPicEvent_OrderBy;
(function (NewPicEvent_OrderBy) {
    NewPicEvent_OrderBy["Id"] = "id";
    NewPicEvent_OrderBy["TransactionHash"] = "transactionHash";
    NewPicEvent_OrderBy["GasPrice"] = "gasPrice";
    NewPicEvent_OrderBy["GasUsed"] = "gasUsed";
    NewPicEvent_OrderBy["Timestamp"] = "timestamp";
    NewPicEvent_OrderBy["Name"] = "name";
    NewPicEvent_OrderBy["Addresses"] = "addresses";
    NewPicEvent_OrderBy["BlockNumber"] = "blockNumber";
    NewPicEvent_OrderBy["LogIndex"] = "logIndex";
    NewPicEvent_OrderBy["Order"] = "order";
    NewPicEvent_OrderBy["Token"] = "token";
    NewPicEvent_OrderBy["Pic"] = "pic";
    NewPicEvent_OrderBy["Bond"] = "bond";
    NewPicEvent_OrderBy["ExitRate"] = "exitRate";
})(NewPicEvent_OrderBy = exports.NewPicEvent_OrderBy || (exports.NewPicEvent_OrderBy = {}));
/** Defines the order direction, either ascending or descending */
var OrderDirection;
(function (OrderDirection) {
    OrderDirection["Asc"] = "asc";
    OrderDirection["Desc"] = "desc";
})(OrderDirection = exports.OrderDirection || (exports.OrderDirection = {}));
var PppConfigurationChangedEvent_OrderBy;
(function (PppConfigurationChangedEvent_OrderBy) {
    PppConfigurationChangedEvent_OrderBy["Id"] = "id";
    PppConfigurationChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    PppConfigurationChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    PppConfigurationChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    PppConfigurationChangedEvent_OrderBy["Timestamp"] = "timestamp";
    PppConfigurationChangedEvent_OrderBy["Name"] = "name";
    PppConfigurationChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    PppConfigurationChangedEvent_OrderBy["Addresses"] = "addresses";
    PppConfigurationChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    PppConfigurationChangedEvent_OrderBy["LogIndex"] = "logIndex";
    PppConfigurationChangedEvent_OrderBy["Order"] = "order";
    PppConfigurationChangedEvent_OrderBy["Host"] = "host";
    PppConfigurationChangedEvent_OrderBy["SuperToken"] = "superToken";
    PppConfigurationChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    PppConfigurationChangedEvent_OrderBy["LiquidationPeriod"] = "liquidationPeriod";
    PppConfigurationChangedEvent_OrderBy["PatricianPeriod"] = "patricianPeriod";
})(PppConfigurationChangedEvent_OrderBy = exports.PppConfigurationChangedEvent_OrderBy || (exports.PppConfigurationChangedEvent_OrderBy = {}));
var ResolverEntry_OrderBy;
(function (ResolverEntry_OrderBy) {
    ResolverEntry_OrderBy["Id"] = "id";
    ResolverEntry_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    ResolverEntry_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    ResolverEntry_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    ResolverEntry_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    ResolverEntry_OrderBy["TargetAddress"] = "targetAddress";
    ResolverEntry_OrderBy["IsToken"] = "isToken";
    ResolverEntry_OrderBy["IsListed"] = "isListed";
    ResolverEntry_OrderBy["SetEvents"] = "setEvents";
})(ResolverEntry_OrderBy = exports.ResolverEntry_OrderBy || (exports.ResolverEntry_OrderBy = {}));
var RewardAddressChangedEvent_OrderBy;
(function (RewardAddressChangedEvent_OrderBy) {
    RewardAddressChangedEvent_OrderBy["Id"] = "id";
    RewardAddressChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    RewardAddressChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    RewardAddressChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    RewardAddressChangedEvent_OrderBy["Timestamp"] = "timestamp";
    RewardAddressChangedEvent_OrderBy["Name"] = "name";
    RewardAddressChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    RewardAddressChangedEvent_OrderBy["Addresses"] = "addresses";
    RewardAddressChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    RewardAddressChangedEvent_OrderBy["LogIndex"] = "logIndex";
    RewardAddressChangedEvent_OrderBy["Order"] = "order";
    RewardAddressChangedEvent_OrderBy["Host"] = "host";
    RewardAddressChangedEvent_OrderBy["SuperToken"] = "superToken";
    RewardAddressChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    RewardAddressChangedEvent_OrderBy["RewardAddress"] = "rewardAddress";
})(RewardAddressChangedEvent_OrderBy = exports.RewardAddressChangedEvent_OrderBy || (exports.RewardAddressChangedEvent_OrderBy = {}));
var RoleAdminChangedEvent_OrderBy;
(function (RoleAdminChangedEvent_OrderBy) {
    RoleAdminChangedEvent_OrderBy["Id"] = "id";
    RoleAdminChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    RoleAdminChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    RoleAdminChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    RoleAdminChangedEvent_OrderBy["Timestamp"] = "timestamp";
    RoleAdminChangedEvent_OrderBy["Name"] = "name";
    RoleAdminChangedEvent_OrderBy["Addresses"] = "addresses";
    RoleAdminChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    RoleAdminChangedEvent_OrderBy["LogIndex"] = "logIndex";
    RoleAdminChangedEvent_OrderBy["Order"] = "order";
    RoleAdminChangedEvent_OrderBy["Role"] = "role";
    RoleAdminChangedEvent_OrderBy["PreviousAdminRole"] = "previousAdminRole";
    RoleAdminChangedEvent_OrderBy["NewAdminRole"] = "newAdminRole";
})(RoleAdminChangedEvent_OrderBy = exports.RoleAdminChangedEvent_OrderBy || (exports.RoleAdminChangedEvent_OrderBy = {}));
var RoleGrantedEvent_OrderBy;
(function (RoleGrantedEvent_OrderBy) {
    RoleGrantedEvent_OrderBy["Id"] = "id";
    RoleGrantedEvent_OrderBy["TransactionHash"] = "transactionHash";
    RoleGrantedEvent_OrderBy["GasPrice"] = "gasPrice";
    RoleGrantedEvent_OrderBy["GasUsed"] = "gasUsed";
    RoleGrantedEvent_OrderBy["Timestamp"] = "timestamp";
    RoleGrantedEvent_OrderBy["Name"] = "name";
    RoleGrantedEvent_OrderBy["Addresses"] = "addresses";
    RoleGrantedEvent_OrderBy["BlockNumber"] = "blockNumber";
    RoleGrantedEvent_OrderBy["LogIndex"] = "logIndex";
    RoleGrantedEvent_OrderBy["Order"] = "order";
    RoleGrantedEvent_OrderBy["Role"] = "role";
    RoleGrantedEvent_OrderBy["Account"] = "account";
    RoleGrantedEvent_OrderBy["Sender"] = "sender";
})(RoleGrantedEvent_OrderBy = exports.RoleGrantedEvent_OrderBy || (exports.RoleGrantedEvent_OrderBy = {}));
var RoleRevokedEvent_OrderBy;
(function (RoleRevokedEvent_OrderBy) {
    RoleRevokedEvent_OrderBy["Id"] = "id";
    RoleRevokedEvent_OrderBy["TransactionHash"] = "transactionHash";
    RoleRevokedEvent_OrderBy["GasPrice"] = "gasPrice";
    RoleRevokedEvent_OrderBy["GasUsed"] = "gasUsed";
    RoleRevokedEvent_OrderBy["Timestamp"] = "timestamp";
    RoleRevokedEvent_OrderBy["Name"] = "name";
    RoleRevokedEvent_OrderBy["Addresses"] = "addresses";
    RoleRevokedEvent_OrderBy["BlockNumber"] = "blockNumber";
    RoleRevokedEvent_OrderBy["LogIndex"] = "logIndex";
    RoleRevokedEvent_OrderBy["Order"] = "order";
    RoleRevokedEvent_OrderBy["Role"] = "role";
    RoleRevokedEvent_OrderBy["Account"] = "account";
    RoleRevokedEvent_OrderBy["Sender"] = "sender";
})(RoleRevokedEvent_OrderBy = exports.RoleRevokedEvent_OrderBy || (exports.RoleRevokedEvent_OrderBy = {}));
var SfMeta_OrderBy;
(function (SfMeta_OrderBy) {
    SfMeta_OrderBy["Id"] = "id";
    SfMeta_OrderBy["Timestamp"] = "timestamp";
    SfMeta_OrderBy["BlockNumber"] = "blockNumber";
    SfMeta_OrderBy["Configuration"] = "configuration";
    SfMeta_OrderBy["Branch"] = "branch";
})(SfMeta_OrderBy = exports.SfMeta_OrderBy || (exports.SfMeta_OrderBy = {}));
var SentEvent_OrderBy;
(function (SentEvent_OrderBy) {
    SentEvent_OrderBy["Id"] = "id";
    SentEvent_OrderBy["TransactionHash"] = "transactionHash";
    SentEvent_OrderBy["GasPrice"] = "gasPrice";
    SentEvent_OrderBy["GasUsed"] = "gasUsed";
    SentEvent_OrderBy["Timestamp"] = "timestamp";
    SentEvent_OrderBy["Name"] = "name";
    SentEvent_OrderBy["Addresses"] = "addresses";
    SentEvent_OrderBy["BlockNumber"] = "blockNumber";
    SentEvent_OrderBy["LogIndex"] = "logIndex";
    SentEvent_OrderBy["Order"] = "order";
    SentEvent_OrderBy["Operator"] = "operator";
    SentEvent_OrderBy["From"] = "from";
    SentEvent_OrderBy["To"] = "to";
    SentEvent_OrderBy["Token"] = "token";
    SentEvent_OrderBy["Amount"] = "amount";
    SentEvent_OrderBy["Data"] = "data";
    SentEvent_OrderBy["OperatorData"] = "operatorData";
})(SentEvent_OrderBy = exports.SentEvent_OrderBy || (exports.SentEvent_OrderBy = {}));
var SetEvent_OrderBy;
(function (SetEvent_OrderBy) {
    SetEvent_OrderBy["Id"] = "id";
    SetEvent_OrderBy["TransactionHash"] = "transactionHash";
    SetEvent_OrderBy["GasPrice"] = "gasPrice";
    SetEvent_OrderBy["GasUsed"] = "gasUsed";
    SetEvent_OrderBy["Timestamp"] = "timestamp";
    SetEvent_OrderBy["Name"] = "name";
    SetEvent_OrderBy["Addresses"] = "addresses";
    SetEvent_OrderBy["BlockNumber"] = "blockNumber";
    SetEvent_OrderBy["LogIndex"] = "logIndex";
    SetEvent_OrderBy["Order"] = "order";
    SetEvent_OrderBy["HashedName"] = "hashedName";
    SetEvent_OrderBy["Target"] = "target";
    SetEvent_OrderBy["ResolverEntry"] = "resolverEntry";
    SetEvent_OrderBy["ResolverEntry__id"] = "resolverEntry__id";
    SetEvent_OrderBy["ResolverEntry__createdAtTimestamp"] = "resolverEntry__createdAtTimestamp";
    SetEvent_OrderBy["ResolverEntry__createdAtBlockNumber"] = "resolverEntry__createdAtBlockNumber";
    SetEvent_OrderBy["ResolverEntry__updatedAtTimestamp"] = "resolverEntry__updatedAtTimestamp";
    SetEvent_OrderBy["ResolverEntry__updatedAtBlockNumber"] = "resolverEntry__updatedAtBlockNumber";
    SetEvent_OrderBy["ResolverEntry__targetAddress"] = "resolverEntry__targetAddress";
    SetEvent_OrderBy["ResolverEntry__isToken"] = "resolverEntry__isToken";
    SetEvent_OrderBy["ResolverEntry__isListed"] = "resolverEntry__isListed";
})(SetEvent_OrderBy = exports.SetEvent_OrderBy || (exports.SetEvent_OrderBy = {}));
var StreamPeriod_OrderBy;
(function (StreamPeriod_OrderBy) {
    StreamPeriod_OrderBy["Id"] = "id";
    StreamPeriod_OrderBy["Stream"] = "stream";
    StreamPeriod_OrderBy["Stream__id"] = "stream__id";
    StreamPeriod_OrderBy["Stream__createdAtTimestamp"] = "stream__createdAtTimestamp";
    StreamPeriod_OrderBy["Stream__createdAtBlockNumber"] = "stream__createdAtBlockNumber";
    StreamPeriod_OrderBy["Stream__updatedAtTimestamp"] = "stream__updatedAtTimestamp";
    StreamPeriod_OrderBy["Stream__updatedAtBlockNumber"] = "stream__updatedAtBlockNumber";
    StreamPeriod_OrderBy["Stream__currentFlowRate"] = "stream__currentFlowRate";
    StreamPeriod_OrderBy["Stream__deposit"] = "stream__deposit";
    StreamPeriod_OrderBy["Stream__streamedUntilUpdatedAt"] = "stream__streamedUntilUpdatedAt";
    StreamPeriod_OrderBy["Stream__userData"] = "stream__userData";
    StreamPeriod_OrderBy["Sender"] = "sender";
    StreamPeriod_OrderBy["Sender__id"] = "sender__id";
    StreamPeriod_OrderBy["Sender__createdAtTimestamp"] = "sender__createdAtTimestamp";
    StreamPeriod_OrderBy["Sender__createdAtBlockNumber"] = "sender__createdAtBlockNumber";
    StreamPeriod_OrderBy["Sender__updatedAtTimestamp"] = "sender__updatedAtTimestamp";
    StreamPeriod_OrderBy["Sender__updatedAtBlockNumber"] = "sender__updatedAtBlockNumber";
    StreamPeriod_OrderBy["Sender__isSuperApp"] = "sender__isSuperApp";
    StreamPeriod_OrderBy["Receiver"] = "receiver";
    StreamPeriod_OrderBy["Receiver__id"] = "receiver__id";
    StreamPeriod_OrderBy["Receiver__createdAtTimestamp"] = "receiver__createdAtTimestamp";
    StreamPeriod_OrderBy["Receiver__createdAtBlockNumber"] = "receiver__createdAtBlockNumber";
    StreamPeriod_OrderBy["Receiver__updatedAtTimestamp"] = "receiver__updatedAtTimestamp";
    StreamPeriod_OrderBy["Receiver__updatedAtBlockNumber"] = "receiver__updatedAtBlockNumber";
    StreamPeriod_OrderBy["Receiver__isSuperApp"] = "receiver__isSuperApp";
    StreamPeriod_OrderBy["Token"] = "token";
    StreamPeriod_OrderBy["Token__id"] = "token__id";
    StreamPeriod_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    StreamPeriod_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    StreamPeriod_OrderBy["Token__decimals"] = "token__decimals";
    StreamPeriod_OrderBy["Token__name"] = "token__name";
    StreamPeriod_OrderBy["Token__symbol"] = "token__symbol";
    StreamPeriod_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    StreamPeriod_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    StreamPeriod_OrderBy["Token__isListed"] = "token__isListed";
    StreamPeriod_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    StreamPeriod_OrderBy["FlowRate"] = "flowRate";
    StreamPeriod_OrderBy["Deposit"] = "deposit";
    StreamPeriod_OrderBy["StartedAtTimestamp"] = "startedAtTimestamp";
    StreamPeriod_OrderBy["StartedAtBlockNumber"] = "startedAtBlockNumber";
    StreamPeriod_OrderBy["StartedAtEvent"] = "startedAtEvent";
    StreamPeriod_OrderBy["StartedAtEvent__id"] = "startedAtEvent__id";
    StreamPeriod_OrderBy["StartedAtEvent__transactionHash"] = "startedAtEvent__transactionHash";
    StreamPeriod_OrderBy["StartedAtEvent__gasPrice"] = "startedAtEvent__gasPrice";
    StreamPeriod_OrderBy["StartedAtEvent__gasUsed"] = "startedAtEvent__gasUsed";
    StreamPeriod_OrderBy["StartedAtEvent__timestamp"] = "startedAtEvent__timestamp";
    StreamPeriod_OrderBy["StartedAtEvent__name"] = "startedAtEvent__name";
    StreamPeriod_OrderBy["StartedAtEvent__blockNumber"] = "startedAtEvent__blockNumber";
    StreamPeriod_OrderBy["StartedAtEvent__logIndex"] = "startedAtEvent__logIndex";
    StreamPeriod_OrderBy["StartedAtEvent__order"] = "startedAtEvent__order";
    StreamPeriod_OrderBy["StartedAtEvent__token"] = "startedAtEvent__token";
    StreamPeriod_OrderBy["StartedAtEvent__sender"] = "startedAtEvent__sender";
    StreamPeriod_OrderBy["StartedAtEvent__receiver"] = "startedAtEvent__receiver";
    StreamPeriod_OrderBy["StartedAtEvent__flowOperator"] = "startedAtEvent__flowOperator";
    StreamPeriod_OrderBy["StartedAtEvent__flowRate"] = "startedAtEvent__flowRate";
    StreamPeriod_OrderBy["StartedAtEvent__totalSenderFlowRate"] = "startedAtEvent__totalSenderFlowRate";
    StreamPeriod_OrderBy["StartedAtEvent__totalReceiverFlowRate"] = "startedAtEvent__totalReceiverFlowRate";
    StreamPeriod_OrderBy["StartedAtEvent__deposit"] = "startedAtEvent__deposit";
    StreamPeriod_OrderBy["StartedAtEvent__userData"] = "startedAtEvent__userData";
    StreamPeriod_OrderBy["StartedAtEvent__oldFlowRate"] = "startedAtEvent__oldFlowRate";
    StreamPeriod_OrderBy["StartedAtEvent__type"] = "startedAtEvent__type";
    StreamPeriod_OrderBy["StartedAtEvent__totalAmountStreamedUntilTimestamp"] = "startedAtEvent__totalAmountStreamedUntilTimestamp";
    StreamPeriod_OrderBy["StoppedAtTimestamp"] = "stoppedAtTimestamp";
    StreamPeriod_OrderBy["StoppedAtBlockNumber"] = "stoppedAtBlockNumber";
    StreamPeriod_OrderBy["StoppedAtEvent"] = "stoppedAtEvent";
    StreamPeriod_OrderBy["StoppedAtEvent__id"] = "stoppedAtEvent__id";
    StreamPeriod_OrderBy["StoppedAtEvent__transactionHash"] = "stoppedAtEvent__transactionHash";
    StreamPeriod_OrderBy["StoppedAtEvent__gasPrice"] = "stoppedAtEvent__gasPrice";
    StreamPeriod_OrderBy["StoppedAtEvent__gasUsed"] = "stoppedAtEvent__gasUsed";
    StreamPeriod_OrderBy["StoppedAtEvent__timestamp"] = "stoppedAtEvent__timestamp";
    StreamPeriod_OrderBy["StoppedAtEvent__name"] = "stoppedAtEvent__name";
    StreamPeriod_OrderBy["StoppedAtEvent__blockNumber"] = "stoppedAtEvent__blockNumber";
    StreamPeriod_OrderBy["StoppedAtEvent__logIndex"] = "stoppedAtEvent__logIndex";
    StreamPeriod_OrderBy["StoppedAtEvent__order"] = "stoppedAtEvent__order";
    StreamPeriod_OrderBy["StoppedAtEvent__token"] = "stoppedAtEvent__token";
    StreamPeriod_OrderBy["StoppedAtEvent__sender"] = "stoppedAtEvent__sender";
    StreamPeriod_OrderBy["StoppedAtEvent__receiver"] = "stoppedAtEvent__receiver";
    StreamPeriod_OrderBy["StoppedAtEvent__flowOperator"] = "stoppedAtEvent__flowOperator";
    StreamPeriod_OrderBy["StoppedAtEvent__flowRate"] = "stoppedAtEvent__flowRate";
    StreamPeriod_OrderBy["StoppedAtEvent__totalSenderFlowRate"] = "stoppedAtEvent__totalSenderFlowRate";
    StreamPeriod_OrderBy["StoppedAtEvent__totalReceiverFlowRate"] = "stoppedAtEvent__totalReceiverFlowRate";
    StreamPeriod_OrderBy["StoppedAtEvent__deposit"] = "stoppedAtEvent__deposit";
    StreamPeriod_OrderBy["StoppedAtEvent__userData"] = "stoppedAtEvent__userData";
    StreamPeriod_OrderBy["StoppedAtEvent__oldFlowRate"] = "stoppedAtEvent__oldFlowRate";
    StreamPeriod_OrderBy["StoppedAtEvent__type"] = "stoppedAtEvent__type";
    StreamPeriod_OrderBy["StoppedAtEvent__totalAmountStreamedUntilTimestamp"] = "stoppedAtEvent__totalAmountStreamedUntilTimestamp";
    StreamPeriod_OrderBy["TotalAmountStreamed"] = "totalAmountStreamed";
})(StreamPeriod_OrderBy = exports.StreamPeriod_OrderBy || (exports.StreamPeriod_OrderBy = {}));
var StreamRevision_OrderBy;
(function (StreamRevision_OrderBy) {
    StreamRevision_OrderBy["Id"] = "id";
    StreamRevision_OrderBy["RevisionIndex"] = "revisionIndex";
    StreamRevision_OrderBy["PeriodRevisionIndex"] = "periodRevisionIndex";
    StreamRevision_OrderBy["MostRecentStream"] = "mostRecentStream";
    StreamRevision_OrderBy["MostRecentStream__id"] = "mostRecentStream__id";
    StreamRevision_OrderBy["MostRecentStream__createdAtTimestamp"] = "mostRecentStream__createdAtTimestamp";
    StreamRevision_OrderBy["MostRecentStream__createdAtBlockNumber"] = "mostRecentStream__createdAtBlockNumber";
    StreamRevision_OrderBy["MostRecentStream__updatedAtTimestamp"] = "mostRecentStream__updatedAtTimestamp";
    StreamRevision_OrderBy["MostRecentStream__updatedAtBlockNumber"] = "mostRecentStream__updatedAtBlockNumber";
    StreamRevision_OrderBy["MostRecentStream__currentFlowRate"] = "mostRecentStream__currentFlowRate";
    StreamRevision_OrderBy["MostRecentStream__deposit"] = "mostRecentStream__deposit";
    StreamRevision_OrderBy["MostRecentStream__streamedUntilUpdatedAt"] = "mostRecentStream__streamedUntilUpdatedAt";
    StreamRevision_OrderBy["MostRecentStream__userData"] = "mostRecentStream__userData";
})(StreamRevision_OrderBy = exports.StreamRevision_OrderBy || (exports.StreamRevision_OrderBy = {}));
var Stream_OrderBy;
(function (Stream_OrderBy) {
    Stream_OrderBy["Id"] = "id";
    Stream_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    Stream_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    Stream_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    Stream_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    Stream_OrderBy["CurrentFlowRate"] = "currentFlowRate";
    Stream_OrderBy["Deposit"] = "deposit";
    Stream_OrderBy["StreamedUntilUpdatedAt"] = "streamedUntilUpdatedAt";
    Stream_OrderBy["Token"] = "token";
    Stream_OrderBy["Token__id"] = "token__id";
    Stream_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    Stream_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    Stream_OrderBy["Token__decimals"] = "token__decimals";
    Stream_OrderBy["Token__name"] = "token__name";
    Stream_OrderBy["Token__symbol"] = "token__symbol";
    Stream_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    Stream_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    Stream_OrderBy["Token__isListed"] = "token__isListed";
    Stream_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    Stream_OrderBy["Sender"] = "sender";
    Stream_OrderBy["Sender__id"] = "sender__id";
    Stream_OrderBy["Sender__createdAtTimestamp"] = "sender__createdAtTimestamp";
    Stream_OrderBy["Sender__createdAtBlockNumber"] = "sender__createdAtBlockNumber";
    Stream_OrderBy["Sender__updatedAtTimestamp"] = "sender__updatedAtTimestamp";
    Stream_OrderBy["Sender__updatedAtBlockNumber"] = "sender__updatedAtBlockNumber";
    Stream_OrderBy["Sender__isSuperApp"] = "sender__isSuperApp";
    Stream_OrderBy["Receiver"] = "receiver";
    Stream_OrderBy["Receiver__id"] = "receiver__id";
    Stream_OrderBy["Receiver__createdAtTimestamp"] = "receiver__createdAtTimestamp";
    Stream_OrderBy["Receiver__createdAtBlockNumber"] = "receiver__createdAtBlockNumber";
    Stream_OrderBy["Receiver__updatedAtTimestamp"] = "receiver__updatedAtTimestamp";
    Stream_OrderBy["Receiver__updatedAtBlockNumber"] = "receiver__updatedAtBlockNumber";
    Stream_OrderBy["Receiver__isSuperApp"] = "receiver__isSuperApp";
    Stream_OrderBy["UserData"] = "userData";
    Stream_OrderBy["FlowUpdatedEvents"] = "flowUpdatedEvents";
    Stream_OrderBy["StreamPeriods"] = "streamPeriods";
})(Stream_OrderBy = exports.Stream_OrderBy || (exports.Stream_OrderBy = {}));
var SubscriptionApprovedEvent_OrderBy;
(function (SubscriptionApprovedEvent_OrderBy) {
    SubscriptionApprovedEvent_OrderBy["Id"] = "id";
    SubscriptionApprovedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SubscriptionApprovedEvent_OrderBy["GasPrice"] = "gasPrice";
    SubscriptionApprovedEvent_OrderBy["GasUsed"] = "gasUsed";
    SubscriptionApprovedEvent_OrderBy["Timestamp"] = "timestamp";
    SubscriptionApprovedEvent_OrderBy["Name"] = "name";
    SubscriptionApprovedEvent_OrderBy["Addresses"] = "addresses";
    SubscriptionApprovedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SubscriptionApprovedEvent_OrderBy["LogIndex"] = "logIndex";
    SubscriptionApprovedEvent_OrderBy["Order"] = "order";
    SubscriptionApprovedEvent_OrderBy["Token"] = "token";
    SubscriptionApprovedEvent_OrderBy["Subscriber"] = "subscriber";
    SubscriptionApprovedEvent_OrderBy["Publisher"] = "publisher";
    SubscriptionApprovedEvent_OrderBy["IndexId"] = "indexId";
    SubscriptionApprovedEvent_OrderBy["UserData"] = "userData";
    SubscriptionApprovedEvent_OrderBy["Subscription"] = "subscription";
    SubscriptionApprovedEvent_OrderBy["Subscription__id"] = "subscription__id";
    SubscriptionApprovedEvent_OrderBy["Subscription__createdAtTimestamp"] = "subscription__createdAtTimestamp";
    SubscriptionApprovedEvent_OrderBy["Subscription__createdAtBlockNumber"] = "subscription__createdAtBlockNumber";
    SubscriptionApprovedEvent_OrderBy["Subscription__updatedAtTimestamp"] = "subscription__updatedAtTimestamp";
    SubscriptionApprovedEvent_OrderBy["Subscription__updatedAtBlockNumber"] = "subscription__updatedAtBlockNumber";
    SubscriptionApprovedEvent_OrderBy["Subscription__approved"] = "subscription__approved";
    SubscriptionApprovedEvent_OrderBy["Subscription__units"] = "subscription__units";
    SubscriptionApprovedEvent_OrderBy["Subscription__totalAmountReceivedUntilUpdatedAt"] = "subscription__totalAmountReceivedUntilUpdatedAt";
    SubscriptionApprovedEvent_OrderBy["Subscription__indexValueUntilUpdatedAt"] = "subscription__indexValueUntilUpdatedAt";
})(SubscriptionApprovedEvent_OrderBy = exports.SubscriptionApprovedEvent_OrderBy || (exports.SubscriptionApprovedEvent_OrderBy = {}));
var SubscriptionDistributionClaimedEvent_OrderBy;
(function (SubscriptionDistributionClaimedEvent_OrderBy) {
    SubscriptionDistributionClaimedEvent_OrderBy["Id"] = "id";
    SubscriptionDistributionClaimedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SubscriptionDistributionClaimedEvent_OrderBy["GasPrice"] = "gasPrice";
    SubscriptionDistributionClaimedEvent_OrderBy["GasUsed"] = "gasUsed";
    SubscriptionDistributionClaimedEvent_OrderBy["Timestamp"] = "timestamp";
    SubscriptionDistributionClaimedEvent_OrderBy["Name"] = "name";
    SubscriptionDistributionClaimedEvent_OrderBy["Addresses"] = "addresses";
    SubscriptionDistributionClaimedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SubscriptionDistributionClaimedEvent_OrderBy["LogIndex"] = "logIndex";
    SubscriptionDistributionClaimedEvent_OrderBy["Order"] = "order";
    SubscriptionDistributionClaimedEvent_OrderBy["Token"] = "token";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscriber"] = "subscriber";
    SubscriptionDistributionClaimedEvent_OrderBy["Publisher"] = "publisher";
    SubscriptionDistributionClaimedEvent_OrderBy["IndexId"] = "indexId";
    SubscriptionDistributionClaimedEvent_OrderBy["Amount"] = "amount";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription"] = "subscription";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__id"] = "subscription__id";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__createdAtTimestamp"] = "subscription__createdAtTimestamp";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__createdAtBlockNumber"] = "subscription__createdAtBlockNumber";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__updatedAtTimestamp"] = "subscription__updatedAtTimestamp";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__updatedAtBlockNumber"] = "subscription__updatedAtBlockNumber";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__approved"] = "subscription__approved";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__units"] = "subscription__units";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__totalAmountReceivedUntilUpdatedAt"] = "subscription__totalAmountReceivedUntilUpdatedAt";
    SubscriptionDistributionClaimedEvent_OrderBy["Subscription__indexValueUntilUpdatedAt"] = "subscription__indexValueUntilUpdatedAt";
})(SubscriptionDistributionClaimedEvent_OrderBy = exports.SubscriptionDistributionClaimedEvent_OrderBy || (exports.SubscriptionDistributionClaimedEvent_OrderBy = {}));
var SubscriptionRevokedEvent_OrderBy;
(function (SubscriptionRevokedEvent_OrderBy) {
    SubscriptionRevokedEvent_OrderBy["Id"] = "id";
    SubscriptionRevokedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SubscriptionRevokedEvent_OrderBy["GasPrice"] = "gasPrice";
    SubscriptionRevokedEvent_OrderBy["GasUsed"] = "gasUsed";
    SubscriptionRevokedEvent_OrderBy["Timestamp"] = "timestamp";
    SubscriptionRevokedEvent_OrderBy["Name"] = "name";
    SubscriptionRevokedEvent_OrderBy["Addresses"] = "addresses";
    SubscriptionRevokedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SubscriptionRevokedEvent_OrderBy["LogIndex"] = "logIndex";
    SubscriptionRevokedEvent_OrderBy["Order"] = "order";
    SubscriptionRevokedEvent_OrderBy["Token"] = "token";
    SubscriptionRevokedEvent_OrderBy["Subscriber"] = "subscriber";
    SubscriptionRevokedEvent_OrderBy["Publisher"] = "publisher";
    SubscriptionRevokedEvent_OrderBy["IndexId"] = "indexId";
    SubscriptionRevokedEvent_OrderBy["UserData"] = "userData";
    SubscriptionRevokedEvent_OrderBy["Subscription"] = "subscription";
    SubscriptionRevokedEvent_OrderBy["Subscription__id"] = "subscription__id";
    SubscriptionRevokedEvent_OrderBy["Subscription__createdAtTimestamp"] = "subscription__createdAtTimestamp";
    SubscriptionRevokedEvent_OrderBy["Subscription__createdAtBlockNumber"] = "subscription__createdAtBlockNumber";
    SubscriptionRevokedEvent_OrderBy["Subscription__updatedAtTimestamp"] = "subscription__updatedAtTimestamp";
    SubscriptionRevokedEvent_OrderBy["Subscription__updatedAtBlockNumber"] = "subscription__updatedAtBlockNumber";
    SubscriptionRevokedEvent_OrderBy["Subscription__approved"] = "subscription__approved";
    SubscriptionRevokedEvent_OrderBy["Subscription__units"] = "subscription__units";
    SubscriptionRevokedEvent_OrderBy["Subscription__totalAmountReceivedUntilUpdatedAt"] = "subscription__totalAmountReceivedUntilUpdatedAt";
    SubscriptionRevokedEvent_OrderBy["Subscription__indexValueUntilUpdatedAt"] = "subscription__indexValueUntilUpdatedAt";
})(SubscriptionRevokedEvent_OrderBy = exports.SubscriptionRevokedEvent_OrderBy || (exports.SubscriptionRevokedEvent_OrderBy = {}));
var SubscriptionUnitsUpdatedEvent_OrderBy;
(function (SubscriptionUnitsUpdatedEvent_OrderBy) {
    SubscriptionUnitsUpdatedEvent_OrderBy["Id"] = "id";
    SubscriptionUnitsUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SubscriptionUnitsUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    SubscriptionUnitsUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    SubscriptionUnitsUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    SubscriptionUnitsUpdatedEvent_OrderBy["Name"] = "name";
    SubscriptionUnitsUpdatedEvent_OrderBy["Addresses"] = "addresses";
    SubscriptionUnitsUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SubscriptionUnitsUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    SubscriptionUnitsUpdatedEvent_OrderBy["Order"] = "order";
    SubscriptionUnitsUpdatedEvent_OrderBy["Token"] = "token";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscriber"] = "subscriber";
    SubscriptionUnitsUpdatedEvent_OrderBy["Publisher"] = "publisher";
    SubscriptionUnitsUpdatedEvent_OrderBy["IndexId"] = "indexId";
    SubscriptionUnitsUpdatedEvent_OrderBy["Units"] = "units";
    SubscriptionUnitsUpdatedEvent_OrderBy["UserData"] = "userData";
    SubscriptionUnitsUpdatedEvent_OrderBy["OldUnits"] = "oldUnits";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription"] = "subscription";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__id"] = "subscription__id";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__createdAtTimestamp"] = "subscription__createdAtTimestamp";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__createdAtBlockNumber"] = "subscription__createdAtBlockNumber";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__updatedAtTimestamp"] = "subscription__updatedAtTimestamp";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__updatedAtBlockNumber"] = "subscription__updatedAtBlockNumber";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__approved"] = "subscription__approved";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__units"] = "subscription__units";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__totalAmountReceivedUntilUpdatedAt"] = "subscription__totalAmountReceivedUntilUpdatedAt";
    SubscriptionUnitsUpdatedEvent_OrderBy["Subscription__indexValueUntilUpdatedAt"] = "subscription__indexValueUntilUpdatedAt";
})(SubscriptionUnitsUpdatedEvent_OrderBy = exports.SubscriptionUnitsUpdatedEvent_OrderBy || (exports.SubscriptionUnitsUpdatedEvent_OrderBy = {}));
var SuperTokenCreatedEvent_OrderBy;
(function (SuperTokenCreatedEvent_OrderBy) {
    SuperTokenCreatedEvent_OrderBy["Id"] = "id";
    SuperTokenCreatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SuperTokenCreatedEvent_OrderBy["GasPrice"] = "gasPrice";
    SuperTokenCreatedEvent_OrderBy["GasUsed"] = "gasUsed";
    SuperTokenCreatedEvent_OrderBy["Timestamp"] = "timestamp";
    SuperTokenCreatedEvent_OrderBy["Name"] = "name";
    SuperTokenCreatedEvent_OrderBy["Addresses"] = "addresses";
    SuperTokenCreatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SuperTokenCreatedEvent_OrderBy["LogIndex"] = "logIndex";
    SuperTokenCreatedEvent_OrderBy["Order"] = "order";
    SuperTokenCreatedEvent_OrderBy["Token"] = "token";
})(SuperTokenCreatedEvent_OrderBy = exports.SuperTokenCreatedEvent_OrderBy || (exports.SuperTokenCreatedEvent_OrderBy = {}));
var SuperTokenFactoryUpdatedEvent_OrderBy;
(function (SuperTokenFactoryUpdatedEvent_OrderBy) {
    SuperTokenFactoryUpdatedEvent_OrderBy["Id"] = "id";
    SuperTokenFactoryUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SuperTokenFactoryUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    SuperTokenFactoryUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    SuperTokenFactoryUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    SuperTokenFactoryUpdatedEvent_OrderBy["Name"] = "name";
    SuperTokenFactoryUpdatedEvent_OrderBy["Addresses"] = "addresses";
    SuperTokenFactoryUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SuperTokenFactoryUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    SuperTokenFactoryUpdatedEvent_OrderBy["Order"] = "order";
    SuperTokenFactoryUpdatedEvent_OrderBy["NewFactory"] = "newFactory";
})(SuperTokenFactoryUpdatedEvent_OrderBy = exports.SuperTokenFactoryUpdatedEvent_OrderBy || (exports.SuperTokenFactoryUpdatedEvent_OrderBy = {}));
var SuperTokenLogicCreatedEvent_OrderBy;
(function (SuperTokenLogicCreatedEvent_OrderBy) {
    SuperTokenLogicCreatedEvent_OrderBy["Id"] = "id";
    SuperTokenLogicCreatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SuperTokenLogicCreatedEvent_OrderBy["GasPrice"] = "gasPrice";
    SuperTokenLogicCreatedEvent_OrderBy["GasUsed"] = "gasUsed";
    SuperTokenLogicCreatedEvent_OrderBy["Timestamp"] = "timestamp";
    SuperTokenLogicCreatedEvent_OrderBy["Name"] = "name";
    SuperTokenLogicCreatedEvent_OrderBy["Addresses"] = "addresses";
    SuperTokenLogicCreatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SuperTokenLogicCreatedEvent_OrderBy["LogIndex"] = "logIndex";
    SuperTokenLogicCreatedEvent_OrderBy["Order"] = "order";
    SuperTokenLogicCreatedEvent_OrderBy["TokenLogic"] = "tokenLogic";
})(SuperTokenLogicCreatedEvent_OrderBy = exports.SuperTokenLogicCreatedEvent_OrderBy || (exports.SuperTokenLogicCreatedEvent_OrderBy = {}));
var SuperTokenLogicUpdatedEvent_OrderBy;
(function (SuperTokenLogicUpdatedEvent_OrderBy) {
    SuperTokenLogicUpdatedEvent_OrderBy["Id"] = "id";
    SuperTokenLogicUpdatedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SuperTokenLogicUpdatedEvent_OrderBy["GasPrice"] = "gasPrice";
    SuperTokenLogicUpdatedEvent_OrderBy["GasUsed"] = "gasUsed";
    SuperTokenLogicUpdatedEvent_OrderBy["Timestamp"] = "timestamp";
    SuperTokenLogicUpdatedEvent_OrderBy["Name"] = "name";
    SuperTokenLogicUpdatedEvent_OrderBy["Addresses"] = "addresses";
    SuperTokenLogicUpdatedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SuperTokenLogicUpdatedEvent_OrderBy["LogIndex"] = "logIndex";
    SuperTokenLogicUpdatedEvent_OrderBy["Order"] = "order";
    SuperTokenLogicUpdatedEvent_OrderBy["Token"] = "token";
    SuperTokenLogicUpdatedEvent_OrderBy["Code"] = "code";
})(SuperTokenLogicUpdatedEvent_OrderBy = exports.SuperTokenLogicUpdatedEvent_OrderBy || (exports.SuperTokenLogicUpdatedEvent_OrderBy = {}));
var SuperTokenMinimumDepositChangedEvent_OrderBy;
(function (SuperTokenMinimumDepositChangedEvent_OrderBy) {
    SuperTokenMinimumDepositChangedEvent_OrderBy["Id"] = "id";
    SuperTokenMinimumDepositChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    SuperTokenMinimumDepositChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    SuperTokenMinimumDepositChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    SuperTokenMinimumDepositChangedEvent_OrderBy["Timestamp"] = "timestamp";
    SuperTokenMinimumDepositChangedEvent_OrderBy["Name"] = "name";
    SuperTokenMinimumDepositChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    SuperTokenMinimumDepositChangedEvent_OrderBy["Addresses"] = "addresses";
    SuperTokenMinimumDepositChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    SuperTokenMinimumDepositChangedEvent_OrderBy["LogIndex"] = "logIndex";
    SuperTokenMinimumDepositChangedEvent_OrderBy["Order"] = "order";
    SuperTokenMinimumDepositChangedEvent_OrderBy["Host"] = "host";
    SuperTokenMinimumDepositChangedEvent_OrderBy["SuperToken"] = "superToken";
    SuperTokenMinimumDepositChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    SuperTokenMinimumDepositChangedEvent_OrderBy["MinimumDeposit"] = "minimumDeposit";
})(SuperTokenMinimumDepositChangedEvent_OrderBy = exports.SuperTokenMinimumDepositChangedEvent_OrderBy || (exports.SuperTokenMinimumDepositChangedEvent_OrderBy = {}));
var TokenDowngradedEvent_OrderBy;
(function (TokenDowngradedEvent_OrderBy) {
    TokenDowngradedEvent_OrderBy["Id"] = "id";
    TokenDowngradedEvent_OrderBy["Account"] = "account";
    TokenDowngradedEvent_OrderBy["Account__id"] = "account__id";
    TokenDowngradedEvent_OrderBy["Account__createdAtTimestamp"] = "account__createdAtTimestamp";
    TokenDowngradedEvent_OrderBy["Account__createdAtBlockNumber"] = "account__createdAtBlockNumber";
    TokenDowngradedEvent_OrderBy["Account__updatedAtTimestamp"] = "account__updatedAtTimestamp";
    TokenDowngradedEvent_OrderBy["Account__updatedAtBlockNumber"] = "account__updatedAtBlockNumber";
    TokenDowngradedEvent_OrderBy["Account__isSuperApp"] = "account__isSuperApp";
    TokenDowngradedEvent_OrderBy["TransactionHash"] = "transactionHash";
    TokenDowngradedEvent_OrderBy["GasPrice"] = "gasPrice";
    TokenDowngradedEvent_OrderBy["GasUsed"] = "gasUsed";
    TokenDowngradedEvent_OrderBy["Timestamp"] = "timestamp";
    TokenDowngradedEvent_OrderBy["Name"] = "name";
    TokenDowngradedEvent_OrderBy["Addresses"] = "addresses";
    TokenDowngradedEvent_OrderBy["BlockNumber"] = "blockNumber";
    TokenDowngradedEvent_OrderBy["LogIndex"] = "logIndex";
    TokenDowngradedEvent_OrderBy["Order"] = "order";
    TokenDowngradedEvent_OrderBy["Token"] = "token";
    TokenDowngradedEvent_OrderBy["Amount"] = "amount";
})(TokenDowngradedEvent_OrderBy = exports.TokenDowngradedEvent_OrderBy || (exports.TokenDowngradedEvent_OrderBy = {}));
var TokenGovernanceConfig_OrderBy;
(function (TokenGovernanceConfig_OrderBy) {
    TokenGovernanceConfig_OrderBy["Id"] = "id";
    TokenGovernanceConfig_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    TokenGovernanceConfig_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    TokenGovernanceConfig_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    TokenGovernanceConfig_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    TokenGovernanceConfig_OrderBy["IsDefault"] = "isDefault";
    TokenGovernanceConfig_OrderBy["RewardAddress"] = "rewardAddress";
    TokenGovernanceConfig_OrderBy["LiquidationPeriod"] = "liquidationPeriod";
    TokenGovernanceConfig_OrderBy["PatricianPeriod"] = "patricianPeriod";
    TokenGovernanceConfig_OrderBy["MinimumDeposit"] = "minimumDeposit";
    TokenGovernanceConfig_OrderBy["Token"] = "token";
    TokenGovernanceConfig_OrderBy["Token__id"] = "token__id";
    TokenGovernanceConfig_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    TokenGovernanceConfig_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    TokenGovernanceConfig_OrderBy["Token__decimals"] = "token__decimals";
    TokenGovernanceConfig_OrderBy["Token__name"] = "token__name";
    TokenGovernanceConfig_OrderBy["Token__symbol"] = "token__symbol";
    TokenGovernanceConfig_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    TokenGovernanceConfig_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    TokenGovernanceConfig_OrderBy["Token__isListed"] = "token__isListed";
    TokenGovernanceConfig_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
})(TokenGovernanceConfig_OrderBy = exports.TokenGovernanceConfig_OrderBy || (exports.TokenGovernanceConfig_OrderBy = {}));
var TokenStatisticLog_OrderBy;
(function (TokenStatisticLog_OrderBy) {
    TokenStatisticLog_OrderBy["Id"] = "id";
    TokenStatisticLog_OrderBy["Timestamp"] = "timestamp";
    TokenStatisticLog_OrderBy["BlockNumber"] = "blockNumber";
    TokenStatisticLog_OrderBy["TransactionHash"] = "transactionHash";
    TokenStatisticLog_OrderBy["LogIndex"] = "logIndex";
    TokenStatisticLog_OrderBy["Order"] = "order";
    TokenStatisticLog_OrderBy["TriggeredByEventName"] = "triggeredByEventName";
    TokenStatisticLog_OrderBy["TotalNumberOfActiveStreams"] = "totalNumberOfActiveStreams";
    TokenStatisticLog_OrderBy["TotalNumberOfClosedStreams"] = "totalNumberOfClosedStreams";
    TokenStatisticLog_OrderBy["TotalNumberOfIndexes"] = "totalNumberOfIndexes";
    TokenStatisticLog_OrderBy["TotalNumberOfActiveIndexes"] = "totalNumberOfActiveIndexes";
    TokenStatisticLog_OrderBy["TotalSubscriptionsWithUnits"] = "totalSubscriptionsWithUnits";
    TokenStatisticLog_OrderBy["TotalApprovedSubscriptions"] = "totalApprovedSubscriptions";
    TokenStatisticLog_OrderBy["TotalDeposit"] = "totalDeposit";
    TokenStatisticLog_OrderBy["TotalOutflowRate"] = "totalOutflowRate";
    TokenStatisticLog_OrderBy["TotalAmountStreamed"] = "totalAmountStreamed";
    TokenStatisticLog_OrderBy["TotalAmountTransferred"] = "totalAmountTransferred";
    TokenStatisticLog_OrderBy["TotalAmountDistributed"] = "totalAmountDistributed";
    TokenStatisticLog_OrderBy["TotalSupply"] = "totalSupply";
    TokenStatisticLog_OrderBy["Token"] = "token";
    TokenStatisticLog_OrderBy["Token__id"] = "token__id";
    TokenStatisticLog_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    TokenStatisticLog_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    TokenStatisticLog_OrderBy["Token__decimals"] = "token__decimals";
    TokenStatisticLog_OrderBy["Token__name"] = "token__name";
    TokenStatisticLog_OrderBy["Token__symbol"] = "token__symbol";
    TokenStatisticLog_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    TokenStatisticLog_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    TokenStatisticLog_OrderBy["Token__isListed"] = "token__isListed";
    TokenStatisticLog_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    TokenStatisticLog_OrderBy["TokenStatistic"] = "tokenStatistic";
    TokenStatisticLog_OrderBy["TokenStatistic__id"] = "tokenStatistic__id";
    TokenStatisticLog_OrderBy["TokenStatistic__updatedAtTimestamp"] = "tokenStatistic__updatedAtTimestamp";
    TokenStatisticLog_OrderBy["TokenStatistic__updatedAtBlockNumber"] = "tokenStatistic__updatedAtBlockNumber";
    TokenStatisticLog_OrderBy["TokenStatistic__totalNumberOfActiveStreams"] = "tokenStatistic__totalNumberOfActiveStreams";
    TokenStatisticLog_OrderBy["TokenStatistic__totalNumberOfClosedStreams"] = "tokenStatistic__totalNumberOfClosedStreams";
    TokenStatisticLog_OrderBy["TokenStatistic__totalNumberOfIndexes"] = "tokenStatistic__totalNumberOfIndexes";
    TokenStatisticLog_OrderBy["TokenStatistic__totalNumberOfActiveIndexes"] = "tokenStatistic__totalNumberOfActiveIndexes";
    TokenStatisticLog_OrderBy["TokenStatistic__totalSubscriptionsWithUnits"] = "tokenStatistic__totalSubscriptionsWithUnits";
    TokenStatisticLog_OrderBy["TokenStatistic__totalApprovedSubscriptions"] = "tokenStatistic__totalApprovedSubscriptions";
    TokenStatisticLog_OrderBy["TokenStatistic__totalDeposit"] = "tokenStatistic__totalDeposit";
    TokenStatisticLog_OrderBy["TokenStatistic__totalOutflowRate"] = "tokenStatistic__totalOutflowRate";
    TokenStatisticLog_OrderBy["TokenStatistic__totalAmountStreamedUntilUpdatedAt"] = "tokenStatistic__totalAmountStreamedUntilUpdatedAt";
    TokenStatisticLog_OrderBy["TokenStatistic__totalAmountTransferredUntilUpdatedAt"] = "tokenStatistic__totalAmountTransferredUntilUpdatedAt";
    TokenStatisticLog_OrderBy["TokenStatistic__totalAmountDistributedUntilUpdatedAt"] = "tokenStatistic__totalAmountDistributedUntilUpdatedAt";
    TokenStatisticLog_OrderBy["TokenStatistic__totalSupply"] = "tokenStatistic__totalSupply";
})(TokenStatisticLog_OrderBy = exports.TokenStatisticLog_OrderBy || (exports.TokenStatisticLog_OrderBy = {}));
var TokenStatistic_OrderBy;
(function (TokenStatistic_OrderBy) {
    TokenStatistic_OrderBy["Id"] = "id";
    TokenStatistic_OrderBy["UpdatedAtTimestamp"] = "updatedAtTimestamp";
    TokenStatistic_OrderBy["UpdatedAtBlockNumber"] = "updatedAtBlockNumber";
    TokenStatistic_OrderBy["TotalNumberOfActiveStreams"] = "totalNumberOfActiveStreams";
    TokenStatistic_OrderBy["TotalNumberOfClosedStreams"] = "totalNumberOfClosedStreams";
    TokenStatistic_OrderBy["TotalNumberOfIndexes"] = "totalNumberOfIndexes";
    TokenStatistic_OrderBy["TotalNumberOfActiveIndexes"] = "totalNumberOfActiveIndexes";
    TokenStatistic_OrderBy["TotalSubscriptionsWithUnits"] = "totalSubscriptionsWithUnits";
    TokenStatistic_OrderBy["TotalApprovedSubscriptions"] = "totalApprovedSubscriptions";
    TokenStatistic_OrderBy["TotalDeposit"] = "totalDeposit";
    TokenStatistic_OrderBy["TotalOutflowRate"] = "totalOutflowRate";
    TokenStatistic_OrderBy["TotalAmountStreamedUntilUpdatedAt"] = "totalAmountStreamedUntilUpdatedAt";
    TokenStatistic_OrderBy["TotalAmountTransferredUntilUpdatedAt"] = "totalAmountTransferredUntilUpdatedAt";
    TokenStatistic_OrderBy["TotalAmountDistributedUntilUpdatedAt"] = "totalAmountDistributedUntilUpdatedAt";
    TokenStatistic_OrderBy["TotalSupply"] = "totalSupply";
    TokenStatistic_OrderBy["Token"] = "token";
    TokenStatistic_OrderBy["Token__id"] = "token__id";
    TokenStatistic_OrderBy["Token__createdAtTimestamp"] = "token__createdAtTimestamp";
    TokenStatistic_OrderBy["Token__createdAtBlockNumber"] = "token__createdAtBlockNumber";
    TokenStatistic_OrderBy["Token__decimals"] = "token__decimals";
    TokenStatistic_OrderBy["Token__name"] = "token__name";
    TokenStatistic_OrderBy["Token__symbol"] = "token__symbol";
    TokenStatistic_OrderBy["Token__isSuperToken"] = "token__isSuperToken";
    TokenStatistic_OrderBy["Token__isNativeAssetSuperToken"] = "token__isNativeAssetSuperToken";
    TokenStatistic_OrderBy["Token__isListed"] = "token__isListed";
    TokenStatistic_OrderBy["Token__underlyingAddress"] = "token__underlyingAddress";
    TokenStatistic_OrderBy["TokenStatisticLogs"] = "tokenStatisticLogs";
})(TokenStatistic_OrderBy = exports.TokenStatistic_OrderBy || (exports.TokenStatistic_OrderBy = {}));
var TokenUpgradedEvent_OrderBy;
(function (TokenUpgradedEvent_OrderBy) {
    TokenUpgradedEvent_OrderBy["Id"] = "id";
    TokenUpgradedEvent_OrderBy["Account"] = "account";
    TokenUpgradedEvent_OrderBy["Account__id"] = "account__id";
    TokenUpgradedEvent_OrderBy["Account__createdAtTimestamp"] = "account__createdAtTimestamp";
    TokenUpgradedEvent_OrderBy["Account__createdAtBlockNumber"] = "account__createdAtBlockNumber";
    TokenUpgradedEvent_OrderBy["Account__updatedAtTimestamp"] = "account__updatedAtTimestamp";
    TokenUpgradedEvent_OrderBy["Account__updatedAtBlockNumber"] = "account__updatedAtBlockNumber";
    TokenUpgradedEvent_OrderBy["Account__isSuperApp"] = "account__isSuperApp";
    TokenUpgradedEvent_OrderBy["TransactionHash"] = "transactionHash";
    TokenUpgradedEvent_OrderBy["GasPrice"] = "gasPrice";
    TokenUpgradedEvent_OrderBy["GasUsed"] = "gasUsed";
    TokenUpgradedEvent_OrderBy["Timestamp"] = "timestamp";
    TokenUpgradedEvent_OrderBy["Name"] = "name";
    TokenUpgradedEvent_OrderBy["Addresses"] = "addresses";
    TokenUpgradedEvent_OrderBy["BlockNumber"] = "blockNumber";
    TokenUpgradedEvent_OrderBy["LogIndex"] = "logIndex";
    TokenUpgradedEvent_OrderBy["Order"] = "order";
    TokenUpgradedEvent_OrderBy["Token"] = "token";
    TokenUpgradedEvent_OrderBy["Amount"] = "amount";
})(TokenUpgradedEvent_OrderBy = exports.TokenUpgradedEvent_OrderBy || (exports.TokenUpgradedEvent_OrderBy = {}));
var Token_OrderBy;
(function (Token_OrderBy) {
    Token_OrderBy["Id"] = "id";
    Token_OrderBy["CreatedAtTimestamp"] = "createdAtTimestamp";
    Token_OrderBy["CreatedAtBlockNumber"] = "createdAtBlockNumber";
    Token_OrderBy["Decimals"] = "decimals";
    Token_OrderBy["Name"] = "name";
    Token_OrderBy["Symbol"] = "symbol";
    Token_OrderBy["IsSuperToken"] = "isSuperToken";
    Token_OrderBy["IsNativeAssetSuperToken"] = "isNativeAssetSuperToken";
    Token_OrderBy["IsListed"] = "isListed";
    Token_OrderBy["UnderlyingAddress"] = "underlyingAddress";
    Token_OrderBy["UnderlyingToken"] = "underlyingToken";
    Token_OrderBy["UnderlyingToken__id"] = "underlyingToken__id";
    Token_OrderBy["UnderlyingToken__createdAtTimestamp"] = "underlyingToken__createdAtTimestamp";
    Token_OrderBy["UnderlyingToken__createdAtBlockNumber"] = "underlyingToken__createdAtBlockNumber";
    Token_OrderBy["UnderlyingToken__decimals"] = "underlyingToken__decimals";
    Token_OrderBy["UnderlyingToken__name"] = "underlyingToken__name";
    Token_OrderBy["UnderlyingToken__symbol"] = "underlyingToken__symbol";
    Token_OrderBy["UnderlyingToken__isSuperToken"] = "underlyingToken__isSuperToken";
    Token_OrderBy["UnderlyingToken__isNativeAssetSuperToken"] = "underlyingToken__isNativeAssetSuperToken";
    Token_OrderBy["UnderlyingToken__isListed"] = "underlyingToken__isListed";
    Token_OrderBy["UnderlyingToken__underlyingAddress"] = "underlyingToken__underlyingAddress";
    Token_OrderBy["GovernanceConfig"] = "governanceConfig";
    Token_OrderBy["GovernanceConfig__id"] = "governanceConfig__id";
    Token_OrderBy["GovernanceConfig__createdAtTimestamp"] = "governanceConfig__createdAtTimestamp";
    Token_OrderBy["GovernanceConfig__createdAtBlockNumber"] = "governanceConfig__createdAtBlockNumber";
    Token_OrderBy["GovernanceConfig__updatedAtTimestamp"] = "governanceConfig__updatedAtTimestamp";
    Token_OrderBy["GovernanceConfig__updatedAtBlockNumber"] = "governanceConfig__updatedAtBlockNumber";
    Token_OrderBy["GovernanceConfig__isDefault"] = "governanceConfig__isDefault";
    Token_OrderBy["GovernanceConfig__rewardAddress"] = "governanceConfig__rewardAddress";
    Token_OrderBy["GovernanceConfig__liquidationPeriod"] = "governanceConfig__liquidationPeriod";
    Token_OrderBy["GovernanceConfig__patricianPeriod"] = "governanceConfig__patricianPeriod";
    Token_OrderBy["GovernanceConfig__minimumDeposit"] = "governanceConfig__minimumDeposit";
})(Token_OrderBy = exports.Token_OrderBy || (exports.Token_OrderBy = {}));
var TransferEvent_OrderBy;
(function (TransferEvent_OrderBy) {
    TransferEvent_OrderBy["Id"] = "id";
    TransferEvent_OrderBy["TransactionHash"] = "transactionHash";
    TransferEvent_OrderBy["GasPrice"] = "gasPrice";
    TransferEvent_OrderBy["GasUsed"] = "gasUsed";
    TransferEvent_OrderBy["Timestamp"] = "timestamp";
    TransferEvent_OrderBy["Name"] = "name";
    TransferEvent_OrderBy["Addresses"] = "addresses";
    TransferEvent_OrderBy["BlockNumber"] = "blockNumber";
    TransferEvent_OrderBy["LogIndex"] = "logIndex";
    TransferEvent_OrderBy["Order"] = "order";
    TransferEvent_OrderBy["From"] = "from";
    TransferEvent_OrderBy["From__id"] = "from__id";
    TransferEvent_OrderBy["From__createdAtTimestamp"] = "from__createdAtTimestamp";
    TransferEvent_OrderBy["From__createdAtBlockNumber"] = "from__createdAtBlockNumber";
    TransferEvent_OrderBy["From__updatedAtTimestamp"] = "from__updatedAtTimestamp";
    TransferEvent_OrderBy["From__updatedAtBlockNumber"] = "from__updatedAtBlockNumber";
    TransferEvent_OrderBy["From__isSuperApp"] = "from__isSuperApp";
    TransferEvent_OrderBy["To"] = "to";
    TransferEvent_OrderBy["To__id"] = "to__id";
    TransferEvent_OrderBy["To__createdAtTimestamp"] = "to__createdAtTimestamp";
    TransferEvent_OrderBy["To__createdAtBlockNumber"] = "to__createdAtBlockNumber";
    TransferEvent_OrderBy["To__updatedAtTimestamp"] = "to__updatedAtTimestamp";
    TransferEvent_OrderBy["To__updatedAtBlockNumber"] = "to__updatedAtBlockNumber";
    TransferEvent_OrderBy["To__isSuperApp"] = "to__isSuperApp";
    TransferEvent_OrderBy["Value"] = "value";
    TransferEvent_OrderBy["Token"] = "token";
})(TransferEvent_OrderBy = exports.TransferEvent_OrderBy || (exports.TransferEvent_OrderBy = {}));
var TrustedForwarderChangedEvent_OrderBy;
(function (TrustedForwarderChangedEvent_OrderBy) {
    TrustedForwarderChangedEvent_OrderBy["Id"] = "id";
    TrustedForwarderChangedEvent_OrderBy["TransactionHash"] = "transactionHash";
    TrustedForwarderChangedEvent_OrderBy["GasPrice"] = "gasPrice";
    TrustedForwarderChangedEvent_OrderBy["GasUsed"] = "gasUsed";
    TrustedForwarderChangedEvent_OrderBy["Timestamp"] = "timestamp";
    TrustedForwarderChangedEvent_OrderBy["Name"] = "name";
    TrustedForwarderChangedEvent_OrderBy["GovernanceAddress"] = "governanceAddress";
    TrustedForwarderChangedEvent_OrderBy["Addresses"] = "addresses";
    TrustedForwarderChangedEvent_OrderBy["BlockNumber"] = "blockNumber";
    TrustedForwarderChangedEvent_OrderBy["LogIndex"] = "logIndex";
    TrustedForwarderChangedEvent_OrderBy["Order"] = "order";
    TrustedForwarderChangedEvent_OrderBy["Host"] = "host";
    TrustedForwarderChangedEvent_OrderBy["SuperToken"] = "superToken";
    TrustedForwarderChangedEvent_OrderBy["IsKeySet"] = "isKeySet";
    TrustedForwarderChangedEvent_OrderBy["Forwarder"] = "forwarder";
    TrustedForwarderChangedEvent_OrderBy["Enabled"] = "enabled";
})(TrustedForwarderChangedEvent_OrderBy = exports.TrustedForwarderChangedEvent_OrderBy || (exports.TrustedForwarderChangedEvent_OrderBy = {}));
var _SubgraphErrorPolicy_;
(function (_SubgraphErrorPolicy_) {
    /** Data will be returned even if the subgraph has indexing errors */
    _SubgraphErrorPolicy_["Allow"] = "allow";
    /** If the subgraph has indexing errors, data will be omitted. The default. */
    _SubgraphErrorPolicy_["Deny"] = "deny";
})(_SubgraphErrorPolicy_ = exports._SubgraphErrorPolicy_ || (exports._SubgraphErrorPolicy_ = {}));
exports.EventFragmentFragmentDoc = (0, graphql_tag_1.default) `
    fragment EventFragment on FlowUpdatedEvent {
  transactionHash
  blockNumber
  timestamp
  sender
  flowRate
  oldFlowRate
  type
}
    `;
exports.PaymentFragmentFragmentDoc = (0, graphql_tag_1.default) `
    fragment PaymentFragment on SentEvent {
  transactionHash
  blockNumber
  timestamp
  from
  amount
}
    `;
exports.GetSuperFluidEventsDocument = (0, graphql_tag_1.default) `
    query GetSuperFluidEvents($tokenAddress: Bytes!, $to: Bytes!, $reference: Bytes!) {
  untagged: flowUpdatedEvents(
    orderBy: timestamp
    where: {flowRate: "0", userData: "0x", type: 2, token: $tokenAddress, receiver: $to}
  ) {
    ...EventFragment
  }
  flow: flowUpdatedEvents(
    orderBy: timestamp
    where: {userData: $reference, token: $tokenAddress, receiver: $to}
  ) {
    ...EventFragment
  }
  payment: sentEvents(
    orderBy: timestamp
    where: {data: $reference, token: $tokenAddress, to: $to}
  ) {
    ...PaymentFragment
  }
}
    ${exports.EventFragmentFragmentDoc}
${exports.PaymentFragmentFragmentDoc}`;
const defaultWrapper = (action, _operationName) => action();
function getSdk(client, withWrapper = defaultWrapper) {
    return {
        GetSuperFluidEvents(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(exports.GetSuperFluidEventsDocument, variables, Object.assign(Object.assign({}, requestHeaders), wrappedRequestHeaders)), 'GetSuperFluidEvents');
        }
    };
}
exports.getSdk = getSdk;
exports.EventFragment = (0, graphql_tag_1.default) `
    fragment EventFragment on FlowUpdatedEvent {
  transactionHash
  blockNumber
  timestamp
  sender
  flowRate
  oldFlowRate
  type
}
    `;
exports.PaymentFragment = (0, graphql_tag_1.default) `
    fragment PaymentFragment on SentEvent {
  transactionHash
  blockNumber
  timestamp
  from
  amount
}
    `;
exports.GetSuperFluidEvents = (0, graphql_tag_1.default) `
    query GetSuperFluidEvents($tokenAddress: Bytes!, $to: Bytes!, $reference: Bytes!) {
  untagged: flowUpdatedEvents(
    orderBy: timestamp
    where: {flowRate: "0", userData: "0x", type: 2, token: $tokenAddress, receiver: $to}
  ) {
    ...EventFragment
  }
  flow: flowUpdatedEvents(
    orderBy: timestamp
    where: {userData: $reference, token: $tokenAddress, receiver: $to}
  ) {
    ...EventFragment
  }
  payment: sentEvents(
    orderBy: timestamp
    where: {data: $reference, token: $tokenAddress, to: $to}
  ) {
    ...PaymentFragment
  }
}
    ${exports.EventFragment}
${exports.PaymentFragment}`;
